{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"/home/lau/postgap/lib/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import postgap \n",
    "import postgap.Globals\n",
    "import postgap.Integration\n",
    "import sys\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "from postgap.Utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests, sys\n",
    "server = \"http://rest.ensembl.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "postgap.Globals.DATABASES_DIR = '/home/lau/postgap/databases'\n",
    "postgap.Globals.SPECIES='human'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = '1'\n",
    "pairs_f = '../github/data/GTEx_Analysis_v7_eQTL_EVB_linearData_hiC.csv'\n",
    "rs_name = 'rs_id_dbSNP147_GRCh37p13'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(396, 210)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>enhancer_ATF2</th>\n",
       "      <th>enhancer_ATF3</th>\n",
       "      <th>enhancer_BATF</th>\n",
       "      <th>enhancer_BCL11A</th>\n",
       "      <th>enhancer_BCL3</th>\n",
       "      <th>enhancer_BCLAF1</th>\n",
       "      <th>...</th>\n",
       "      <th>promoter_ZNF143</th>\n",
       "      <th>promoter_ZNF384</th>\n",
       "      <th>promoter_ZZZ3</th>\n",
       "      <th>hiC_raw</th>\n",
       "      <th>hiC_kr</th>\n",
       "      <th>hiC_vc</th>\n",
       "      <th>hiC_sq</th>\n",
       "      <th>hiC_krExpected</th>\n",
       "      <th>hiC_vcExpected</th>\n",
       "      <th>hiC_sqExpected</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3576</th>\n",
       "      <td>1_1239339_T_G_b37</td>\n",
       "      <td>ENSG00000127054</td>\n",
       "      <td>GM12878|chr1:1240136-1240200</td>\n",
       "      <td>GM12878|chr1:1309330-1312050</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>492.157911</td>\n",
       "      <td>2.06497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>24.927210</td>\n",
       "      <td>24.219449</td>\n",
       "      <td>28.722428</td>\n",
       "      <td>2.139792</td>\n",
       "      <td>1.709039</td>\n",
       "      <td>2.433242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3577</th>\n",
       "      <td>1_1241800_A_G_b37</td>\n",
       "      <td>ENSG00000127054</td>\n",
       "      <td>GM12878|chr1:1240136-1240200</td>\n",
       "      <td>GM12878|chr1:1309330-1312050</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>492.157911</td>\n",
       "      <td>2.06497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>24.927210</td>\n",
       "      <td>24.219449</td>\n",
       "      <td>28.722428</td>\n",
       "      <td>2.139792</td>\n",
       "      <td>1.709039</td>\n",
       "      <td>2.433242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3578</th>\n",
       "      <td>1_1242707_A_G_b37</td>\n",
       "      <td>ENSG00000127054</td>\n",
       "      <td>GM12878|chr1:1240136-1240200</td>\n",
       "      <td>GM12878|chr1:1309330-1312050</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>492.157911</td>\n",
       "      <td>2.06497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>24.927210</td>\n",
       "      <td>24.219449</td>\n",
       "      <td>28.722428</td>\n",
       "      <td>2.139792</td>\n",
       "      <td>1.709039</td>\n",
       "      <td>2.433242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3579</th>\n",
       "      <td>1_1242707_A_G_b37</td>\n",
       "      <td>ENSG00000127054</td>\n",
       "      <td>GM12878|chr1:1245169-1245389</td>\n",
       "      <td>GM12878|chr1:1309330-1312050</td>\n",
       "      <td>37.873228</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>492.157911</td>\n",
       "      <td>2.06497</td>\n",
       "      <td>0.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>34.794071</td>\n",
       "      <td>33.922299</td>\n",
       "      <td>34.723489</td>\n",
       "      <td>2.801549</td>\n",
       "      <td>2.588801</td>\n",
       "      <td>2.761328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3580</th>\n",
       "      <td>1_1297483_G_A_b37</td>\n",
       "      <td>ENSG00000242485</td>\n",
       "      <td>GM12878|chr1:1294400-1294888</td>\n",
       "      <td>GM12878|chr1:1406163-1407732</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>12.878777</td>\n",
       "      <td>13.690376</td>\n",
       "      <td>11.029565</td>\n",
       "      <td>1.723805</td>\n",
       "      <td>1.792028</td>\n",
       "      <td>1.455436</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 210 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             variant_id          gene_id                   enhancer_id  \\\n",
       "3576  1_1239339_T_G_b37  ENSG00000127054  GM12878|chr1:1240136-1240200   \n",
       "3577  1_1241800_A_G_b37  ENSG00000127054  GM12878|chr1:1240136-1240200   \n",
       "3578  1_1242707_A_G_b37  ENSG00000127054  GM12878|chr1:1240136-1240200   \n",
       "3579  1_1242707_A_G_b37  ENSG00000127054  GM12878|chr1:1245169-1245389   \n",
       "3580  1_1297483_G_A_b37  ENSG00000242485  GM12878|chr1:1294400-1294888   \n",
       "\n",
       "                       promoter_id  enhancer_ATF2  enhancer_ATF3  \\\n",
       "3576  GM12878|chr1:1309330-1312050       0.000000            0.0   \n",
       "3577  GM12878|chr1:1309330-1312050       0.000000            0.0   \n",
       "3578  GM12878|chr1:1309330-1312050       0.000000            0.0   \n",
       "3579  GM12878|chr1:1309330-1312050      37.873228            0.0   \n",
       "3580  GM12878|chr1:1406163-1407732       0.000000            0.0   \n",
       "\n",
       "      enhancer_BATF  enhancer_BCL11A  enhancer_BCL3  enhancer_BCLAF1  ...  \\\n",
       "3576            0.0              0.0            0.0              0.0  ...   \n",
       "3577            0.0              0.0            0.0              0.0  ...   \n",
       "3578            0.0              0.0            0.0              0.0  ...   \n",
       "3579            0.0              0.0            0.0              0.0  ...   \n",
       "3580            0.0              0.0            0.0              0.0  ...   \n",
       "\n",
       "      promoter_ZNF143  promoter_ZNF384  promoter_ZZZ3  hiC_raw     hiC_kr  \\\n",
       "3576       492.157911          2.06497            0.0     23.0  24.927210   \n",
       "3577       492.157911          2.06497            0.0     23.0  24.927210   \n",
       "3578       492.157911          2.06497            0.0     23.0  24.927210   \n",
       "3579       492.157911          2.06497            0.0     24.0  34.794071   \n",
       "3580         0.000000          0.00000            0.0      6.0  12.878777   \n",
       "\n",
       "         hiC_vc     hiC_sq  hiC_krExpected  hiC_vcExpected  hiC_sqExpected  \n",
       "3576  24.219449  28.722428        2.139792        1.709039        2.433242  \n",
       "3577  24.219449  28.722428        2.139792        1.709039        2.433242  \n",
       "3578  24.219449  28.722428        2.139792        1.709039        2.433242  \n",
       "3579  33.922299  34.723489        2.801549        2.588801        2.761328  \n",
       "3580  13.690376  11.029565        1.723805        1.792028        1.455436  \n",
       "\n",
       "[5 rows x 210 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(pairs_f)\n",
    "data = data[data['variant_id'].apply(lambda x: x.split('_')[0]==chromosome)]\n",
    "print(data.shape)\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10478</td>\n",
       "      <td>1_10478_C_G_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs528916756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10480</td>\n",
       "      <td>1_10480_C_A_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10482</td>\n",
       "      <td>1_10482_G_C_b37</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10489</td>\n",
       "      <td>1_10489_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10497</td>\n",
       "      <td>1_10497_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   0      1                2  3  4  5            6\n",
       "0  1  10478  1_10478_C_G_b37  C  G  1  rs528916756\n",
       "1  1  10480  1_10480_C_A_b37  C  A  1            .\n",
       "2  1  10482  1_10482_G_C_b37  G  C  1            .\n",
       "3  1  10489  1_10489_C_T_b37  C  T  1            .\n",
       "4  1  10497  1_10497_C_T_b37  C  T  1            ."
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs = pd.read_csv('../../prototype/GTEx_Analysis_2016-01-15_v7_Chromosome1.lookup_table.txt.gz', sep='\\t',\\\n",
    "                 header=-1)\n",
    "rs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "rs.columns = ['chr','variant_pos','variant_id','ref','alt','num_alt_per_site', 'rs_id_dbSNP147_GRCh37p13']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10478</td>\n",
       "      <td>1_10478_C_G_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs528916756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10480</td>\n",
       "      <td>1_10480_C_A_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10482</td>\n",
       "      <td>1_10482_G_C_b37</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10489</td>\n",
       "      <td>1_10489_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10497</td>\n",
       "      <td>1_10497_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr  variant_pos       variant_id ref alt  num_alt_per_site  \\\n",
       "0   1        10478  1_10478_C_G_b37   C   G                 1   \n",
       "1   1        10480  1_10480_C_A_b37   C   A                 1   \n",
       "2   1        10482  1_10482_G_C_b37   G   C                 1   \n",
       "3   1        10489  1_10489_C_T_b37   C   T                 1   \n",
       "4   1        10497  1_10497_C_T_b37   C   T                 1   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  \n",
       "0              rs528916756  \n",
       "1                        .  \n",
       "2                        .  \n",
       "3                        .  \n",
       "4                        .  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.merge(rs, on=['variant_id'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "rsIDs = data[rs_name].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Extracted 227 SNPs.\n"
     ]
    }
   ],
   "source": [
    "# extract SNP objects\n",
    "snps = postgap.Ensembl_lookup.get_snp_locations(rsIDs)\n",
    "print('Extracted {} SNPs.'.format(len(snps)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get LD matrix\n",
    "rsIDs, matrix = postgap.LD.get_pairwise_ld(snps, 'EUR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(209, 209)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(rsIDs), len(set(rsIDs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.loc[data[rs_name].isin(rsIDs),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "M = {}\n",
    "for i, a in enumerate(rsIDs):\n",
    "    if a not in M: M[a] = {}\n",
    "    for j, b in enumerate(rsIDs):\n",
    "        M[a][b] = matrix[i,j]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['rs_name'] = data[rs_name] + '_' + data['promoter_id'] + data['enhancer_id'] \n",
    "data = data.drop_duplicates(subset='rs_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR!\n",
      "ERROR!\n",
      "(355, 26)\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "n = 0\n",
    "id_beta = {}\n",
    "for i in data.index:\n",
    "    gene_stbl_id = data.loc[i,]['gene_id']\n",
    "    variant_name = data.loc[i,][rs_name]\n",
    "    ext = \"/eqtl/id/homo_sapiens/\"+gene_stbl_id+\"?statistic=beta;variant_name=\"+variant_name+\";tissue=Cells_EBV-transformed_lymphocytes\" \n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        try:\n",
    "            r.raise_for_status()\n",
    "        except:\n",
    "            n += 1\n",
    "    try:\n",
    "        decoded = r.json()\n",
    "    except:\n",
    "        n += 1\n",
    "    if \"error\" not in repr(decoded) and decoded != []: \n",
    "        if (gene_stbl_id, variant_name) not in id_beta:\n",
    "            id_beta[(gene_stbl_id, variant_name)] = decoded[0]['value']\n",
    "        else:\n",
    "            if decoded[0]['value'] != id_beta[(gene_stbl_id, variant_name)]:\n",
    "                print('ERROR!')\n",
    "        c += 1\n",
    "print(c, n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "ib = {}\n",
    "keys = tuple(id_beta.keys())\n",
    "ib['gene_id'], ib[rs_name] = zip(*keys)\n",
    "ib['beta'] = tuple(id_beta[k] for k in keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(355, 218)"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = data.merge(pd.DataFrame(ib), right_on=['gene_id', rs_name], left_on=['gene_id', rs_name])\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "E = []\n",
    "for a in data[rs_name]:\n",
    "    e = []\n",
    "    for b in data[rs_name]:\n",
    "        e.append(M[a][b])\n",
    "    E.append(e)\n",
    "    \n",
    "E = np.array(E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>min_pval_nominal</th>\n",
       "      <th>pval_beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_1511837_C_G_b37</td>\n",
       "      <td>ENSG00000225972.1</td>\n",
       "      <td>947395</td>\n",
       "      <td>7</td>\n",
       "      <td>8</td>\n",
       "      <td>0.034188</td>\n",
       "      <td>1.104660e-05</td>\n",
       "      <td>-1.152090</td>\n",
       "      <td>0.248181</td>\n",
       "      <td>0.000018</td>\n",
       "      <td>1.104660e-05</td>\n",
       "      <td>0.015085</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_710088_C_A_b37</td>\n",
       "      <td>ENSG00000224956.5</td>\n",
       "      <td>46561</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0.021368</td>\n",
       "      <td>2.210340e-07</td>\n",
       "      <td>-1.401700</td>\n",
       "      <td>0.250950</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>6.974990e-09</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_710195_A_G_b37</td>\n",
       "      <td>ENSG00000224956.5</td>\n",
       "      <td>46668</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>0.025641</td>\n",
       "      <td>1.608950e-08</td>\n",
       "      <td>-1.447660</td>\n",
       "      <td>0.234342</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>6.974990e-09</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_712515_T_TC_b37</td>\n",
       "      <td>ENSG00000224956.5</td>\n",
       "      <td>48988</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>0.080952</td>\n",
       "      <td>1.213410e-07</td>\n",
       "      <td>-0.927573</td>\n",
       "      <td>0.162069</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>6.974990e-09</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_713092_G_A_b37</td>\n",
       "      <td>ENSG00000224956.5</td>\n",
       "      <td>49565</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>0.021368</td>\n",
       "      <td>2.210340e-07</td>\n",
       "      <td>-1.401700</td>\n",
       "      <td>0.250950</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>6.974990e-09</td>\n",
       "      <td>0.000016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id            gene_id  tss_distance  ma_samples  ma_count  \\\n",
       "0  1_1511837_C_G_b37  ENSG00000225972.1        947395           7         8   \n",
       "1   1_710088_C_A_b37  ENSG00000224956.5         46561           4         5   \n",
       "2   1_710195_A_G_b37  ENSG00000224956.5         46668           5         6   \n",
       "3  1_712515_T_TC_b37  ENSG00000224956.5         48988          15        17   \n",
       "4   1_713092_G_A_b37  ENSG00000224956.5         49565           4         5   \n",
       "\n",
       "        maf  pval_nominal     slope  slope_se  pval_nominal_threshold  \\\n",
       "0  0.034188  1.104660e-05 -1.152090  0.248181                0.000018   \n",
       "1  0.021368  2.210340e-07 -1.401700  0.250950                0.000014   \n",
       "2  0.025641  1.608950e-08 -1.447660  0.234342                0.000014   \n",
       "3  0.080952  1.213410e-07 -0.927573  0.162069                0.000014   \n",
       "4  0.021368  2.210340e-07 -1.401700  0.250950                0.000014   \n",
       "\n",
       "   min_pval_nominal  pval_beta  \n",
       "0      1.104660e-05   0.015085  \n",
       "1      6.974990e-09   0.000016  \n",
       "2      6.974990e-09   0.000016  \n",
       "3      6.974990e-09   0.000016  \n",
       "4      6.974990e-09   0.000016  "
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "evb = pd.read_csv('/mnt/c/Users/laura/Documents/EBI_internship/TFG/github/data/GTEx_Analysis_v7_eQTL/Cells_EBV-transformed_lymphocytes.v7.signif_variant_gene_pairs.txt.gz', sep='\\t')\n",
    "evb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "evb['gene_id'] = evb.gene_id.apply(lambda x: x.split('.')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.merge(evb, on=['variant_id', 'gene_id'], how='inner')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_z = data[['rs_name','chr','variant_pos','ref', 'alt', 'maf', 'beta', 'slope_se']]\n",
    "data_z.columns = ['rsid', 'chromosome','position','allele1', 'allele2', 'maf', 'beta', 'se']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((355, 8), (355, 355))"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_z.shape, E.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_z.to_csv('../data/chromosome'+chromosome+'_to_run_finemap.z', index=False, sep=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt('../data/chromosome'+chromosome+'_to_run_finemap.ld', E, delimiter=' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "# write the master:\n",
    "headers = 'z;ld;snp;config;cred;log;n_samples\\n'\n",
    "names   = 'data/chromosome'+chromosome+'_to_run_finemap.z;data/chromosome'+chromosome+'_to_run_finemap.ld;\\\n",
    "data/chromosome'+chromosome+'_to_run_finemap.snp;data/chromosome'+chromosome+'_to_run_finemap.config;data/chromosome'+chromosome+'_to_run_finemap.cred;data/chromosome'+chromosome+'_to_run_finemap.log;'+str(data_z.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/master', 'w') as f:\n",
    "    f.write(headers+names)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### To run fineMap.\n",
    "You need to download the Github repo and follow the instructions there. \n",
    "Then add the path of the program to your path. \n",
    "Also, note that the verision is v1.3.1!                                                     "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "!finemap_v1.3.1_x86_64 --sss --in-files data/master --dataset 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lau/.local/lib/python2.7/site-packages/ipykernel_launcher.py:1: FutureWarning: read_table is deprecated, use read_csv instead.\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "      <th>log10bf</th>\n",
       "      <th>group</th>\n",
       "      <th>corr_group</th>\n",
       "      <th>prob_group</th>\n",
       "      <th>log10bf_group</th>\n",
       "      <th>mean</th>\n",
       "      <th>sd</th>\n",
       "      <th>mean_incl</th>\n",
       "      <th>sd_incl</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>rs1242331_GM12878|chr1:53017504-53019678GM1287...</td>\n",
       "      <td>1</td>\n",
       "      <td>53392908</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.276316</td>\n",
       "      <td>0.611292</td>\n",
       "      <td>0.075897</td>\n",
       "      <td>8.05426</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>12.35050</td>\n",
       "      <td>299</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.3505</td>\n",
       "      <td>0.646919</td>\n",
       "      <td>0.028543</td>\n",
       "      <td>0.646919</td>\n",
       "      <td>0.028543</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>rs3768543_GM12878|chr1:226068902-226071144GM12...</td>\n",
       "      <td>1</td>\n",
       "      <td>226253926</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.256410</td>\n",
       "      <td>0.898322</td>\n",
       "      <td>0.097062</td>\n",
       "      <td>9.25512</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>12.35050</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.3505</td>\n",
       "      <td>0.832843</td>\n",
       "      <td>0.029309</td>\n",
       "      <td>0.832843</td>\n",
       "      <td>0.029309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>rs12039069_GM12878|chr1:33430151-33430865GM128...</td>\n",
       "      <td>1</td>\n",
       "      <td>33854013</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.295652</td>\n",
       "      <td>0.735142</td>\n",
       "      <td>0.086737</td>\n",
       "      <td>8.47556</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>12.35050</td>\n",
       "      <td>260</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.3505</td>\n",
       "      <td>0.701754</td>\n",
       "      <td>0.031670</td>\n",
       "      <td>0.701754</td>\n",
       "      <td>0.031670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>rs11580071_GM12878|chr1:161037430-161040320GM1...</td>\n",
       "      <td>1</td>\n",
       "      <td>160951832</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.094017</td>\n",
       "      <td>1.014930</td>\n",
       "      <td>0.143916</td>\n",
       "      <td>7.05226</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>9.59909</td>\n",
       "      <td>197</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.3505</td>\n",
       "      <td>0.815989</td>\n",
       "      <td>0.044897</td>\n",
       "      <td>0.815989</td>\n",
       "      <td>0.044897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>rs2039590_GM12878|chr1:150737600-150738647GM12...</td>\n",
       "      <td>1</td>\n",
       "      <td>150813315</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.465812</td>\n",
       "      <td>-0.484168</td>\n",
       "      <td>0.061026</td>\n",
       "      <td>-7.93379</td>\n",
       "      <td>0.61039</td>\n",
       "      <td>2.54558</td>\n",
       "      <td>90</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12.3505</td>\n",
       "      <td>-0.345363</td>\n",
       "      <td>0.276615</td>\n",
       "      <td>-0.565807</td>\n",
       "      <td>0.025042</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    rsid  chromosome  \\\n",
       "index                                                                  \n",
       "299    rs1242331_GM12878|chr1:53017504-53019678GM1287...           1   \n",
       "211    rs3768543_GM12878|chr1:226068902-226071144GM12...           1   \n",
       "260    rs12039069_GM12878|chr1:33430151-33430865GM128...           1   \n",
       "197    rs11580071_GM12878|chr1:161037430-161040320GM1...           1   \n",
       "90     rs2039590_GM12878|chr1:150737600-150738647GM12...           1   \n",
       "\n",
       "        position allele1 allele2       maf      beta        se        z  \\\n",
       "index                                                                     \n",
       "299     53392908       A       G  0.276316  0.611292  0.075897  8.05426   \n",
       "211    226253926       T       C  0.256410  0.898322  0.097062  9.25512   \n",
       "260     33854013       C       T  0.295652  0.735142  0.086737  8.47556   \n",
       "197    160951832       G       A  0.094017  1.014930  0.143916  7.05226   \n",
       "90     150813315       G       A  0.465812 -0.484168  0.061026 -7.93379   \n",
       "\n",
       "          prob   log10bf  group  corr_group  prob_group  log10bf_group  \\\n",
       "index                                                                    \n",
       "299    1.00000  12.35050    299           1           1        12.3505   \n",
       "211    1.00000  12.35050    211           1           1        12.3505   \n",
       "260    1.00000  12.35050    260           1           1        12.3505   \n",
       "197    1.00000   9.59909    197           1           1        12.3505   \n",
       "90     0.61039   2.54558     90           1           1        12.3505   \n",
       "\n",
       "           mean        sd  mean_incl   sd_incl  \n",
       "index                                           \n",
       "299    0.646919  0.028543   0.646919  0.028543  \n",
       "211    0.832843  0.029309   0.832843  0.029309  \n",
       "260    0.701754  0.031670   0.701754  0.031670  \n",
       "197    0.815989  0.044897   0.815989  0.044897  \n",
       "90    -0.345363  0.276615  -0.565807  0.025042  "
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "snp = pd.read_table('../data/chromosome1_to_run_finemap.snp', sep=' ', index_col=0)\n",
    "snp.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.merge(snp, right_on='rsid', left_on='rs_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "data.to_csv('../github/data/GTEx_Analysis_v7_eQTL_EVB_linearData_hiC_z.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15rc1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
