{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pybedtools \n",
    "import sys\n",
    "sys.path.append('../../prototype/notebooks/')\n",
    "import methodName\n",
    "from sklearn.ensemble import ExtraTreesRegressor\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import requests, sys\n",
    "server = \"http://rest.ensembl.org\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtex_f = '../../prototype/GTEx_Analysis_v7_eQTL/Cells_EBV-transformed_lymphocytes.v7.signif_variant_gene_pairs.txt.gz'\n",
    "hi_c_folder = '../../prototype/GM12878_diploid_intrachromosomal_contact_matrices/maternal/5kb_resolution_intrachromosomal/chr8/MAPQGE10/'\n",
    "methylation_f = '../../prototype/methylation.bed.gz'\n",
    "cage_f = '../../prototype/cage.bed.gz'\n",
    "peaks_f = '../../prototype/peaks.bed.gz'\n",
    "enhancers_f = '../../prototype/notebooks/list_of_enhancers_state.csv'\n",
    "promoters_f = '../../prototype/notebooks/list_of_promoters_state.csv'\n",
    "genes_f = '../../prototype/genes.bed'\n",
    "rs_f = '../../prototype/rs_chromosome_8.tab'\n",
    "rs_name = 'rs_id_dbSNP147_GRCh37p13'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# List of Pairs for chromosome 8"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>min_pval_nominal</th>\n",
       "      <th>pval_beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>138112</th>\n",
       "      <td>8_636099_G_A_b37</td>\n",
       "      <td>ENSG00000254207.1</td>\n",
       "      <td>-39778</td>\n",
       "      <td>60</td>\n",
       "      <td>73</td>\n",
       "      <td>0.311966</td>\n",
       "      <td>1.445880e-06</td>\n",
       "      <td>0.643478</td>\n",
       "      <td>0.125106</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>8.274150e-20</td>\n",
       "      <td>4.033910e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138113</th>\n",
       "      <td>8_636101_A_C_b37</td>\n",
       "      <td>ENSG00000254207.1</td>\n",
       "      <td>-39776</td>\n",
       "      <td>45</td>\n",
       "      <td>51</td>\n",
       "      <td>0.217949</td>\n",
       "      <td>2.900780e-07</td>\n",
       "      <td>0.733659</td>\n",
       "      <td>0.132846</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>8.274150e-20</td>\n",
       "      <td>4.033910e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138114</th>\n",
       "      <td>8_637069_C_G_b37</td>\n",
       "      <td>ENSG00000254207.1</td>\n",
       "      <td>-38808</td>\n",
       "      <td>49</td>\n",
       "      <td>57</td>\n",
       "      <td>0.243590</td>\n",
       "      <td>2.610690e-07</td>\n",
       "      <td>0.706680</td>\n",
       "      <td>0.127397</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>8.274150e-20</td>\n",
       "      <td>4.033910e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138115</th>\n",
       "      <td>8_638503_G_C_b37</td>\n",
       "      <td>ENSG00000254207.1</td>\n",
       "      <td>-37374</td>\n",
       "      <td>50</td>\n",
       "      <td>61</td>\n",
       "      <td>0.260684</td>\n",
       "      <td>1.685160e-07</td>\n",
       "      <td>0.701395</td>\n",
       "      <td>0.124183</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>8.274150e-20</td>\n",
       "      <td>4.033910e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138116</th>\n",
       "      <td>8_640522_G_A_b37</td>\n",
       "      <td>ENSG00000254207.1</td>\n",
       "      <td>-35355</td>\n",
       "      <td>49</td>\n",
       "      <td>58</td>\n",
       "      <td>0.247863</td>\n",
       "      <td>6.014460e-08</td>\n",
       "      <td>0.752441</td>\n",
       "      <td>0.127906</td>\n",
       "      <td>0.00001</td>\n",
       "      <td>8.274150e-20</td>\n",
       "      <td>4.033910e-15</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              variant_id            gene_id  tss_distance  ma_samples  \\\n",
       "138112  8_636099_G_A_b37  ENSG00000254207.1        -39778          60   \n",
       "138113  8_636101_A_C_b37  ENSG00000254207.1        -39776          45   \n",
       "138114  8_637069_C_G_b37  ENSG00000254207.1        -38808          49   \n",
       "138115  8_638503_G_C_b37  ENSG00000254207.1        -37374          50   \n",
       "138116  8_640522_G_A_b37  ENSG00000254207.1        -35355          49   \n",
       "\n",
       "        ma_count       maf  pval_nominal     slope  slope_se  \\\n",
       "138112        73  0.311966  1.445880e-06  0.643478  0.125106   \n",
       "138113        51  0.217949  2.900780e-07  0.733659  0.132846   \n",
       "138114        57  0.243590  2.610690e-07  0.706680  0.127397   \n",
       "138115        61  0.260684  1.685160e-07  0.701395  0.124183   \n",
       "138116        58  0.247863  6.014460e-08  0.752441  0.127906   \n",
       "\n",
       "        pval_nominal_threshold  min_pval_nominal     pval_beta  \n",
       "138112                 0.00001      8.274150e-20  4.033910e-15  \n",
       "138113                 0.00001      8.274150e-20  4.033910e-15  \n",
       "138114                 0.00001      8.274150e-20  4.033910e-15  \n",
       "138115                 0.00001      8.274150e-20  4.033910e-15  \n",
       "138116                 0.00001      8.274150e-20  4.033910e-15  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read eQTL \n",
    "eQTL = pd.read_csv(gtex_f, sep='\\t')\n",
    "eQTL = eQTL.loc[eQTL['variant_id'].apply(methodName.variantId2chrNum)=='8',]\n",
    "#eQTL['gene_id'] = eQTL['gene_id'].apply(methodName.geneIdVersion2geneId)\n",
    "eQTL.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(('8_1710632_T_C_b37', 'ENSG00000253982'),\n",
       " ('8_1710977_A_C_b37', 'ENSG00000253982'),\n",
       " ('8_1712158_AG_A_b37', 'ENSG00000253982'),\n",
       " ('8_2064932_A_G_b37', 'ENSG00000036448'),\n",
       " ('8_2065430_A_T_b37', 'ENSG00000036448'),\n",
       " ('8_2068801_T_C_b37', 'ENSG00000036448'),\n",
       " ('8_2069598_GAC_G_b37', 'ENSG00000036448'),\n",
       " ('8_2070062_T_C_b37', 'ENSG00000036448'),\n",
       " ('8_2071002_T_C_b37', 'ENSG00000036448'),\n",
       " ('8_2074042_A_G_b37', 'ENSG00000036448'))"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_vg = tuple(zip(eQTL['variant_id'], eQTL['gene_id'].apply(methodName.geneIdVersion2geneId)))\n",
    "pairs_vg[100:110]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get variant id to enhancers dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lau/.local/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  return pandas.read_table(self.fn, *args, **kwargs)\n"
     ]
    }
   ],
   "source": [
    "# Variant bed table\n",
    "variants = pybedtools.BedTool.from_dataframe(pd.DataFrame({\n",
    "    'variant_chr': eQTL['variant_id'].apply(methodName.variantId2chr),\n",
    "    'variant_start': eQTL['variant_id'].apply(methodName.variantId2pos),\n",
    "    'variant_end': eQTL['variant_id'].apply(methodName.variantId2end),\n",
    "    'variant_id': eQTL['variant_id']}))\n",
    "\n",
    "# Enhancer bed table\n",
    "enhancers = pybedtools.BedTool(enhancers_f)\n",
    "\n",
    "# Intersection\n",
    "variants_enhancers_names = ['variant_chr', 'variant_start', 'variant_end', 'variant_id', \\\n",
    "                           'enhancer_chr', 'enhancer_start', 'enhancer_end', 'enhancer_state', 'enhancer_id']\n",
    "variants_enhancers = variants.intersect(enhancers, wa=True, wb=True).to_dataframe(names=variants_enhancers_names)\n",
    "\n",
    "# Save it as a dictionary for the future\n",
    "variantID__enhancersID = {}\n",
    "for i in variants_enhancers.index:\n",
    "    v = variants_enhancers.loc[i,]['variant_id']\n",
    "    e = variants_enhancers.loc[i,]['enhancer_id']\n",
    "    if v not in variantID__enhancersID:\n",
    "        variantID__enhancersID[v] = [e]\n",
    "    else:\n",
    "        variantID__enhancersID[v].append(e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get gene to promoter id dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lau/.local/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  return pandas.read_table(self.fn, *args, **kwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(49767, 14)\n"
     ]
    }
   ],
   "source": [
    "# Read genes table\n",
    "genes = pybedtools.BedTool(genes_f)\n",
    "\n",
    "# Read promoters table\n",
    "promoters = pybedtools.BedTool(promoters_f)\n",
    "\n",
    "# Intersection\n",
    "genes_promoter_names = ['gene_chr', 'gene_start', 'gene_end', 'gene_id', 'smth1', 'strand', 'annotation', 'type', 'smth2', \\\n",
    "                       'promoter_chr', 'promoter_start', 'promoter_end', 'promoter_state', 'promoter_id']\n",
    "genes_promoters = genes.intersect(promoters, wa=True, wb=True).to_dataframe(names = genes_promoter_names)\n",
    "print(genes_promoters.shape)\n",
    "\n",
    "# Save a dictionary for the future\n",
    "geneID__promotersID = {}\n",
    "for i in genes_promoters.index:\n",
    "    g = genes_promoters.loc[i,]['gene_id']\n",
    "    p = genes_promoters.loc[i,]['promoter_id']\n",
    "    if g not in geneID__promotersID:\n",
    "        geneID__promotersID[g] = [p]\n",
    "    else:\n",
    "        geneID__promotersID[g].append(p)\n",
    "        \n",
    "genes_promoters = genes_promoters[['gene_chr', 'gene_start', 'gene_end', 'gene_id', \\\n",
    "                                   'promoter_chr', 'promoter_start', 'promoter_end', 'promoter_state', 'promoter_id']]\n",
    "\n",
    "# Filter to get only chromosome 8\n",
    "genes_chromosome = set(eQTL['gene_id'])\n",
    "geneID__promotersID  = {g:p for g, p in geneID__promotersID.items() if g in set(genes_chromosome)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7952 257 106\n"
     ]
    }
   ],
   "source": [
    "print(len(pairs_vg), len(variantID__enhancersID), len(geneID__promotersID))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "329\n"
     ]
    }
   ],
   "source": [
    "# Get promoter enhancer table\n",
    "variantID = []\n",
    "geneID = []\n",
    "enhancerID = []\n",
    "promoterID = []\n",
    "\n",
    "c = 0\n",
    "for v, g in pairs_vg:\n",
    "    if v in variantID__enhancersID and g in geneID__promotersID:\n",
    "        promoters = geneID__promotersID[g]\n",
    "        enhancers = variantID__enhancersID[v]\n",
    "\n",
    "        pairs_ep = tuple((enhancers[i], promoters[u]) for i in range(len(enhancers)) for u in range(len(promoters)))\n",
    "        \n",
    "        for e, p in pairs_ep:\n",
    "            variantID.append(v)\n",
    "            geneID.append(g)\n",
    "            enhancerID.append(e)\n",
    "            promoterID.append(p)\n",
    "\n",
    "            c += 1\n",
    "print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df = pd.DataFrame({\n",
    "    'variant_id': variantID,\n",
    "    'gene_id': geneID,\n",
    "    'enhancer_id': enhancerID,\n",
    "    'promoter_id': promoterID})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id          gene_id           enhancer_id  \\\n",
       "0  8_6262831_A_T_b37  ENSG00000246089  chr8:6262801-6263000   \n",
       "1  8_8258448_C_G_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "2  8_8258455_G_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "3  8_8258712_T_C_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "4  8_8259117_T_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "\n",
       "            promoter_id  \n",
       "0  chr8:6405000-6408201  \n",
       "1  chr8:8226600-8230601  \n",
       "2  chr8:8226600-8230601  \n",
       "3  chr8:8226600-8230601  \n",
       "4  chr8:8226600-8230601  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>23845</td>\n",
       "      <td>8_23845_G_GT_b37</td>\n",
       "      <td>G</td>\n",
       "      <td>GT</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>44764</td>\n",
       "      <td>8_44764_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>rs544896538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>44811</td>\n",
       "      <td>8_44811_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8</td>\n",
       "      <td>46114</td>\n",
       "      <td>8_46114_C_T_b37</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>46125</td>\n",
       "      <td>8_46125_T_G_b37</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chr  variant_pos        variant_id ref alt  num_alt_per_site  \\\n",
       "0    8        23845  8_23845_G_GT_b37   G  GT                 1   \n",
       "1    8        44764   8_44764_C_T_b37   C   T                 2   \n",
       "2    8        44811   8_44811_C_T_b37   C   T                 1   \n",
       "3    8        46114   8_46114_C_T_b37   C   T                 2   \n",
       "4    8        46125   8_46125_T_G_b37   T   G                 2   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  \n",
       "0                        .  \n",
       "1              rs544896538  \n",
       "2                        .  \n",
       "3                        .  \n",
       "4                        .  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs = pd.read_csv(rs_f, sep='\\t')\n",
    "rs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "      <td>8</td>\n",
       "      <td>6262831</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>rs725438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258448</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258455</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2979150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2976929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8259117</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920991</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id          gene_id           enhancer_id  \\\n",
       "0  8_6262831_A_T_b37  ENSG00000246089  chr8:6262801-6263000   \n",
       "1  8_8258448_C_G_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "2  8_8258455_G_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "3  8_8258712_T_C_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "4  8_8259117_T_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "\n",
       "            promoter_id  chr  variant_pos ref alt  num_alt_per_site  \\\n",
       "0  chr8:6405000-6408201    8      6262831   A   T                 1   \n",
       "1  chr8:8226600-8230601    8      8258448   C   G                 1   \n",
       "2  chr8:8226600-8230601    8      8258455   G   A                 1   \n",
       "3  chr8:8226600-8230601    8      8258712   T   C                 1   \n",
       "4  chr8:8226600-8230601    8      8259117   T   A                 1   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  \n",
       "0                 rs725438  \n",
       "1                rs2920992  \n",
       "2                rs2979150  \n",
       "3                rs2976929  \n",
       "4                rs2920991  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs_name = 'rs_id_dbSNP147_GRCh37p13'\n",
    "pairs_df = pairs_df.merge(rs, right_on='variant_id', left_on='variant_id')\n",
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(329, 10)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(329, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "      <th>tss_distance</th>\n",
       "      <th>ma_samples</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>min_pval_nominal</th>\n",
       "      <th>pval_beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "      <td>8</td>\n",
       "      <td>6262831</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>rs725438</td>\n",
       "      <td>-1832</td>\n",
       "      <td>60</td>\n",
       "      <td>75</td>\n",
       "      <td>0.323276</td>\n",
       "      <td>1.283440e-06</td>\n",
       "      <td>-0.613816</td>\n",
       "      <td>0.118679</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>1.773260e-08</td>\n",
       "      <td>9.862700e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258448</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920992</td>\n",
       "      <td>174312</td>\n",
       "      <td>85</td>\n",
       "      <td>112</td>\n",
       "      <td>0.478632</td>\n",
       "      <td>7.157010e-06</td>\n",
       "      <td>-0.543850</td>\n",
       "      <td>0.114459</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258455</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2979150</td>\n",
       "      <td>174319</td>\n",
       "      <td>85</td>\n",
       "      <td>112</td>\n",
       "      <td>0.478632</td>\n",
       "      <td>7.157010e-06</td>\n",
       "      <td>-0.543850</td>\n",
       "      <td>0.114459</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2976929</td>\n",
       "      <td>174576</td>\n",
       "      <td>85</td>\n",
       "      <td>112</td>\n",
       "      <td>0.478632</td>\n",
       "      <td>7.157010e-06</td>\n",
       "      <td>-0.543850</td>\n",
       "      <td>0.114459</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8259117</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920991</td>\n",
       "      <td>174981</td>\n",
       "      <td>81</td>\n",
       "      <td>107</td>\n",
       "      <td>0.490826</td>\n",
       "      <td>6.822610e-07</td>\n",
       "      <td>-0.616851</td>\n",
       "      <td>0.115898</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id          gene_id           enhancer_id  \\\n",
       "0  8_6262831_A_T_b37  ENSG00000246089  chr8:6262801-6263000   \n",
       "1  8_8258448_C_G_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "2  8_8258455_G_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "3  8_8258712_T_C_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "4  8_8259117_T_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "\n",
       "            promoter_id  chr  variant_pos ref alt  num_alt_per_site  \\\n",
       "0  chr8:6405000-6408201    8      6262831   A   T                 1   \n",
       "1  chr8:8226600-8230601    8      8258448   C   G                 1   \n",
       "2  chr8:8226600-8230601    8      8258455   G   A                 1   \n",
       "3  chr8:8226600-8230601    8      8258712   T   C                 1   \n",
       "4  chr8:8226600-8230601    8      8259117   T   A                 1   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  tss_distance  ma_samples  ma_count       maf  \\\n",
       "0                 rs725438         -1832          60        75  0.323276   \n",
       "1                rs2920992        174312          85       112  0.478632   \n",
       "2                rs2979150        174319          85       112  0.478632   \n",
       "3                rs2976929        174576          85       112  0.478632   \n",
       "4                rs2920991        174981          81       107  0.490826   \n",
       "\n",
       "   pval_nominal     slope  slope_se  pval_nominal_threshold  min_pval_nominal  \\\n",
       "0  1.283440e-06 -0.613816  0.118679                0.000005      1.773260e-08   \n",
       "1  7.157010e-06 -0.543850  0.114459                0.000019      7.250160e-11   \n",
       "2  7.157010e-06 -0.543850  0.114459                0.000019      7.250160e-11   \n",
       "3  7.157010e-06 -0.543850  0.114459                0.000019      7.250160e-11   \n",
       "4  6.822610e-07 -0.616851  0.115898                0.000019      7.250160e-11   \n",
       "\n",
       "      pval_beta  \n",
       "0  9.862700e-05  \n",
       "1  3.037940e-07  \n",
       "2  3.037940e-07  \n",
       "3  3.037940e-07  \n",
       "4  3.037940e-07  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df = pairs_df.merge(eQTL, right_on=['variant_id', 'gene_id'], left_on=['variant_id', 'gene_id'])\n",
    "print(pairs_df.shape)\n",
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding linear data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "enhancers = pybedtools.BedTool(enhancers_f)\n",
    "promoters = pybedtools.BedTool(promoters_f)\n",
    "peaks = pybedtools.BedTool(peaks_f)\n",
    "methylation = pybedtools.BedTool(methylation_f)\n",
    "cage = pybedtools.BedTool(cage_f)\n",
    "linear = peaks.cat(*[methylation, cage], postmerge=False).sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalizeByLen(row):\n",
    "    to_norm = row[1:]\n",
    "    start, end = row[0].split(':')[1].split('-')\n",
    "    l = int(end) - int(start)\n",
    "    row[1:] = row[1:] / l\n",
    "    return row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lau/.local/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  return pandas.read_table(self.fn, *args, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>enhancer_.</th>\n",
       "      <th>enhancer_ATF2</th>\n",
       "      <th>enhancer_ATF3</th>\n",
       "      <th>enhancer_BATF</th>\n",
       "      <th>enhancer_BCL11A</th>\n",
       "      <th>enhancer_BCL3</th>\n",
       "      <th>enhancer_BCLAF1</th>\n",
       "      <th>enhancer_BHLHE40</th>\n",
       "      <th>enhancer_BRCA1</th>\n",
       "      <th>...</th>\n",
       "      <th>enhancer_USF1</th>\n",
       "      <th>enhancer_USF2</th>\n",
       "      <th>enhancer_WRNIP1</th>\n",
       "      <th>enhancer_YY1</th>\n",
       "      <th>enhancer_ZBTB33</th>\n",
       "      <th>enhancer_ZEB1</th>\n",
       "      <th>enhancer_ZNF143</th>\n",
       "      <th>enhancer_ZNF274</th>\n",
       "      <th>enhancer_ZNF384</th>\n",
       "      <th>enhancer_ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>126101</th>\n",
       "      <td>chrX:99711201-99712400</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126102</th>\n",
       "      <td>chrX:99716001-99716800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126103</th>\n",
       "      <td>chrX:99725801-99726000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126104</th>\n",
       "      <td>chrX:99726401-99727000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126105</th>\n",
       "      <td>chrX:99737001-99737200</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   enhancer_id  enhancer_.  enhancer_ATF2  enhancer_ATF3  \\\n",
       "126101  chrX:99711201-99712400         0.0            0.0            0.0   \n",
       "126102  chrX:99716001-99716800         0.0            0.0            0.0   \n",
       "126103  chrX:99725801-99726000         0.0            0.0            0.0   \n",
       "126104  chrX:99726401-99727000         0.0            0.0            0.0   \n",
       "126105  chrX:99737001-99737200         0.0            0.0            0.0   \n",
       "\n",
       "        enhancer_BATF  enhancer_BCL11A  enhancer_BCL3  enhancer_BCLAF1  \\\n",
       "126101            0.0              0.0            0.0              0.0   \n",
       "126102            0.0              0.0            0.0              0.0   \n",
       "126103            0.0              0.0            0.0              0.0   \n",
       "126104            0.0              0.0            0.0              0.0   \n",
       "126105            0.0              0.0            0.0              0.0   \n",
       "\n",
       "        enhancer_BHLHE40  enhancer_BRCA1  ...  enhancer_USF1  enhancer_USF2  \\\n",
       "126101               0.0             0.0  ...            0.0            0.0   \n",
       "126102               0.0             0.0  ...            0.0            0.0   \n",
       "126103               0.0             0.0  ...            0.0            0.0   \n",
       "126104               0.0             0.0  ...            0.0            0.0   \n",
       "126105               0.0             0.0  ...            0.0            0.0   \n",
       "\n",
       "        enhancer_WRNIP1  enhancer_YY1  enhancer_ZBTB33  enhancer_ZEB1  \\\n",
       "126101              0.0           0.0              0.0            0.0   \n",
       "126102              0.0           0.0              0.0            0.0   \n",
       "126103              0.0           0.0              0.0            0.0   \n",
       "126104              0.0           0.0              0.0            0.0   \n",
       "126105              0.0           0.0              0.0            0.0   \n",
       "\n",
       "        enhancer_ZNF143  enhancer_ZNF274  enhancer_ZNF384  enhancer_ZZZ3  \n",
       "126101              0.0              0.0              0.0            0.0  \n",
       "126102              0.0              0.0              0.0            0.0  \n",
       "126103              0.0              0.0              0.0            0.0  \n",
       "126104              0.0              0.0              0.0            0.0  \n",
       "126105              0.0              0.0              0.0            0.0  \n",
       "\n",
       "[5 rows x 102 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "enhancer_names = ['enhancer_chr', 'enhancer_start', 'enhancert_end', 'enhancer_state', 'enhancer_id']\n",
    "peak_names = ['peak_chr', 'peak_start', 'peak_end', 'peak_name', 'peak_value']\n",
    "enhancer_linear = enhancers.intersect(linear, loj=True, wa=True, wb=True).to_dataframe(names = enhancer_names + peak_names)\n",
    "notDrop = ['enhancer_id', 'peak_name', 'peak_value']\n",
    "enhancer_linear = enhancer_linear[notDrop]\n",
    "enhancer_linear = enhancer_linear.groupby(['enhancer_id', 'peak_name']).sum().reset_index()\n",
    "enhancer_linear = enhancer_linear.pivot_table(index='enhancer_id', columns='peak_name', values='peak_value').reset_index().fillna(0)\n",
    "enhancer_linear.columns = ['enhancer_id'] + ['enhancer_' + i for i in enhancer_linear.columns[1:]]\n",
    "enhancer_linear.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lau/.local/lib/python3.6/site-packages/pybedtools/bedtool.py:3439: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\\t'.\n",
      "  return pandas.read_table(self.fn, *args, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>promoter_.</th>\n",
       "      <th>promoter_ATF2</th>\n",
       "      <th>promoter_ATF3</th>\n",
       "      <th>promoter_BATF</th>\n",
       "      <th>promoter_BCL11A</th>\n",
       "      <th>promoter_BCL3</th>\n",
       "      <th>promoter_BCLAF1</th>\n",
       "      <th>promoter_BHLHE40</th>\n",
       "      <th>promoter_BRCA1</th>\n",
       "      <th>...</th>\n",
       "      <th>promoter_USF1</th>\n",
       "      <th>promoter_USF2</th>\n",
       "      <th>promoter_WRNIP1</th>\n",
       "      <th>promoter_YY1</th>\n",
       "      <th>promoter_ZBTB33</th>\n",
       "      <th>promoter_ZEB1</th>\n",
       "      <th>promoter_ZNF143</th>\n",
       "      <th>promoter_ZNF274</th>\n",
       "      <th>promoter_ZNF384</th>\n",
       "      <th>promoter_ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10:100007000-100011201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:100115000-100115401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10:100183200-100187601</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10:100227800-100230401</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>23.381434</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10:100266000-100268800</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 102 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 promoter_id  promoter_.  promoter_ATF2  promoter_ATF3  \\\n",
       "0  chr10:100007000-100011201         0.0            0.0            0.0   \n",
       "1  chr10:100115000-100115401         0.0            0.0            0.0   \n",
       "2  chr10:100183200-100187601         0.0            0.0            0.0   \n",
       "3  chr10:100227800-100230401         0.0            0.0            0.0   \n",
       "4  chr10:100266000-100268800         0.0            0.0            0.0   \n",
       "\n",
       "   promoter_BATF  promoter_BCL11A  promoter_BCL3  promoter_BCLAF1  \\\n",
       "0            0.0              0.0            0.0              0.0   \n",
       "1            0.0              0.0            0.0              0.0   \n",
       "2            0.0              0.0            0.0              0.0   \n",
       "3            0.0              0.0            0.0              0.0   \n",
       "4            0.0              0.0            0.0              0.0   \n",
       "\n",
       "   promoter_BHLHE40  promoter_BRCA1  ...  promoter_USF1  promoter_USF2  \\\n",
       "0               0.0             0.0  ...            0.0            0.0   \n",
       "1               0.0             0.0  ...            0.0            0.0   \n",
       "2               0.0             0.0  ...            0.0            0.0   \n",
       "3               0.0             0.0  ...            0.0            0.0   \n",
       "4               0.0             0.0  ...            0.0            0.0   \n",
       "\n",
       "   promoter_WRNIP1  promoter_YY1  promoter_ZBTB33  promoter_ZEB1  \\\n",
       "0              0.0           0.0              0.0       0.000000   \n",
       "1              0.0           0.0              0.0       0.000000   \n",
       "2              0.0           0.0              0.0       0.000000   \n",
       "3              0.0           0.0              0.0      23.381434   \n",
       "4              0.0           0.0              0.0       0.000000   \n",
       "\n",
       "   promoter_ZNF143  promoter_ZNF274  promoter_ZNF384  promoter_ZZZ3  \n",
       "0              0.0              0.0              0.0            0.0  \n",
       "1              0.0              0.0              0.0            0.0  \n",
       "2              0.0              0.0              0.0            0.0  \n",
       "3              0.0              0.0              0.0            0.0  \n",
       "4              0.0              0.0              0.0            0.0  \n",
       "\n",
       "[5 rows x 102 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak_names = ['peak_chr', 'peak_start', 'peak_end', 'peak_name', 'peak_value']\n",
    "promoter_names = ['promoter_chr', 'promoter_start', 'promoter_end', 'promoter_state', 'promoter_id']\n",
    "promoter_linear = promoters.intersect(linear, loj=True, wa=True, wb=True).to_dataframe(names = promoter_names + peak_names)\n",
    "notDrop = ['promoter_id', 'peak_name', 'peak_value']\n",
    "promoter_linear = promoter_linear[notDrop]\n",
    "promoter_linear = promoter_linear.groupby(['promoter_id', 'peak_name']).sum().reset_index()\n",
    "promoter_linear = promoter_linear.pivot_table(index='promoter_id', columns='peak_name', values='peak_value').reset_index().fillna(0)\n",
    "promoter_linear.columns = ['promoter_id'] + ['promoter_' + i for i in promoter_linear.columns[1:]]\n",
    "promoter_linear.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding Hi-C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome = '8'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosome__contacts = {}\n",
    "chromosome__contacts[chromosome] = methodName.ContactMatrix(hi_c_folder, chrm=chromosome, resl=5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The raw observed matrix file has 4610383 lines.\n"
     ]
    }
   ],
   "source": [
    "contactMatrix = chromosome__contacts[chromosome].rawMatrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df['enhancer_bin'] = pairs_df['enhancer_id'].apply(lambda x: int(x.split(':')[1].split('-')[0]))\\\n",
    "                            .apply(methodName.position2matrixBin)\n",
    "\n",
    "pairs_df['promoter_bin'] = pairs_df['promoter_id'].apply(lambda x: int(x.split(':')[1].split('-')[0]))\\\n",
    "                            .apply(methodName.position2matrixBin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "hi_c_dict = {'promoter_id':[], 'enhancer_id':[], 'hi_c':[], 'variant_id':[], 'gene_id':[]}\n",
    "for i in pairs_df.index:\n",
    "    pb = pairs_df.loc[i,]['promoter_bin']\n",
    "    eb = pairs_df.loc[i,]['enhancer_bin']\n",
    "    p = pairs_df.loc[i,]['promoter_id']\n",
    "    e = pairs_df.loc[i,]['enhancer_id']\n",
    "    v = pairs_df.loc[i,]['variant_id']\n",
    "    g = pairs_df.loc[i,]['gene_id']\n",
    "    x, y = sorted([pb, eb])\n",
    "    if x in contactMatrix and y in contactMatrix[x]:\n",
    "        hi_c_dict['hi_c'].append(contactMatrix[x][y])\n",
    "    else:\n",
    "        hi_c_dict['hi_c'].append(0)\n",
    "    hi_c_dict['variant_id'].append(v)\n",
    "    hi_c_dict['promoter_id'].append(p)\n",
    "    hi_c_dict['enhancer_id'].append(e)\n",
    "    hi_c_dict['gene_id'].append(g)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding eQTL from other tissues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues_names = ['Adipose_Subcutaneous',\n",
    " 'Adipose_Visceral_Omentum',\n",
    " 'Adrenal_Gland',\n",
    " 'Artery_Aorta',\n",
    " 'Artery_Coronary',\n",
    " 'Artery_Tibial',\n",
    " 'Brain_Amygdala',\n",
    " 'Brain_Anterior_cingulate_cortex_BA24',\n",
    " 'Brain_Caudate_basal_ganglia',\n",
    " 'Brain_Cerebellar_Hemisphere',\n",
    " 'Brain_Cerebellum',\n",
    " 'Brain_Cortex',\n",
    " 'Brain_Frontal_Cortex_BA9',\n",
    " 'Brain_Hippocampus',\n",
    " 'Brain_Hypothalamus',\n",
    " 'Brain_Nucleus_accumbens_basal_ganglia',\n",
    " 'Brain_Putamen_basal_ganglia',\n",
    " 'Brain_Spinal_cord_cervical_c-1',\n",
    " 'Brain_Substantia_nigra',\n",
    " 'Breast_Mammary_Tissue',\n",
    " 'Cells_Transformed_fibroblasts',\n",
    " 'Colon_Sigmoid',\n",
    " 'Colon_Transverse',\n",
    " 'Esophagus_Gastroesophageal_Junction',\n",
    " 'Esophagus_Mucosa',\n",
    " 'Esophagus_Muscularis',\n",
    " 'Heart_Atrial_Appendage',\n",
    " 'Heart_Left_Ventricle',\n",
    " 'Liver',\n",
    " 'Lung',\n",
    " 'Minor_Salivary_Gland',\n",
    " 'Muscle_Skeletal',\n",
    " 'Nerve_Tibial',\n",
    " 'Ovary',\n",
    " 'Pancreas',\n",
    " 'Pituitary',\n",
    " 'Prostate',\n",
    " 'Skin_Not_Sun_Exposed_Suprapubic',\n",
    " 'Skin_Sun_Exposed_Lower_leg',\n",
    " 'Small_Intestine_Terminal_Ileum',\n",
    " 'Spleen',\n",
    " 'Stomach',\n",
    " 'Testis',\n",
    " 'Thyroid',\n",
    " 'Uterus',\n",
    " 'Vagina',\n",
    " 'Whole_Blood']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12668 1834\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "n = 0\n",
    "interaction__tissues_pval = {}\n",
    "for i in pairs_df.index:\n",
    "    gene_stbl_id = pairs_df.loc[i,]['gene_id']\n",
    "    variant_name = pairs_df.loc[i,][rs_name]\n",
    "    for tissue in tissues_names:\n",
    "        ext = \"/eqtl/id/homo_sapiens/\"+gene_stbl_id+\"?statistic=p-value;variant_name=\"+variant_name+\";tissue=\" + tissue\n",
    "        r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "        if not r.ok:\n",
    "            try:\n",
    "                r.raise_for_status()\n",
    "            except:\n",
    "                n += 1\n",
    "        try:\n",
    "            decoded = r.json()\n",
    "        except:\n",
    "            n += 1\n",
    "        if \"error\" not in repr(decoded) and len(repr(decoded)) > 2: \n",
    "            if (gene_stbl_id, variant_name) not in interaction__tissues_pval:\n",
    "                interaction__tissues_pval[(gene_stbl_id, variant_name)] = {}\n",
    "                interaction__tissues_pval[(gene_stbl_id, variant_name)][tissue] = decoded[0]['value']\n",
    "            else:\n",
    "                interaction__tissues_pval[(gene_stbl_id, variant_name)][tissue] = decoded[0]['value']\n",
    "            c += 1\n",
    "print(c, n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "      <th>Adipose_Subcutaneous</th>\n",
       "      <th>Adipose_Visceral_Omentum</th>\n",
       "      <th>Adrenal_Gland</th>\n",
       "      <th>Artery_Aorta</th>\n",
       "      <th>Artery_Coronary</th>\n",
       "      <th>Artery_Tibial</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>Brain_Anterior_cingulate_cortex_BA24</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole_Blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>rs725438</td>\n",
       "      <td>1.043269e-21</td>\n",
       "      <td>6.992728e-11</td>\n",
       "      <td>1.086137e-16</td>\n",
       "      <td>2.721382e-17</td>\n",
       "      <td>7.937750e-09</td>\n",
       "      <td>2.789501e-14</td>\n",
       "      <td>1</td>\n",
       "      <td>0.003194</td>\n",
       "      <td>...</td>\n",
       "      <td>6.476060e-10</td>\n",
       "      <td>4.340546e-15</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>3.087270e-10</td>\n",
       "      <td>2.229960e-08</td>\n",
       "      <td>3.699122e-25</td>\n",
       "      <td>2.288615e-13</td>\n",
       "      <td>0.043879</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>9.007175e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2920992</td>\n",
       "      <td>2.691379e-01</td>\n",
       "      <td>3.598285e-01</td>\n",
       "      <td>2.005095e-01</td>\n",
       "      <td>3.509573e-01</td>\n",
       "      <td>2.239074e-02</td>\n",
       "      <td>2.079237e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>...</td>\n",
       "      <td>1.634199e-01</td>\n",
       "      <td>1.208264e-01</td>\n",
       "      <td>0.744984</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.831117e-03</td>\n",
       "      <td>2.916586e-09</td>\n",
       "      <td>8.773499e-04</td>\n",
       "      <td>0.004506</td>\n",
       "      <td>0.437609</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2979150</td>\n",
       "      <td>2.811242e-01</td>\n",
       "      <td>3.600600e-01</td>\n",
       "      <td>2.037117e-01</td>\n",
       "      <td>3.421716e-01</td>\n",
       "      <td>2.260352e-02</td>\n",
       "      <td>2.303030e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000356</td>\n",
       "      <td>...</td>\n",
       "      <td>1.636915e-01</td>\n",
       "      <td>1.190036e-01</td>\n",
       "      <td>0.746993</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.853348e-03</td>\n",
       "      <td>2.886361e-09</td>\n",
       "      <td>9.617443e-04</td>\n",
       "      <td>0.003875</td>\n",
       "      <td>0.441878</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2976929</td>\n",
       "      <td>2.671713e-01</td>\n",
       "      <td>3.598285e-01</td>\n",
       "      <td>2.003048e-01</td>\n",
       "      <td>3.478835e-01</td>\n",
       "      <td>2.228504e-02</td>\n",
       "      <td>2.089616e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>...</td>\n",
       "      <td>1.634199e-01</td>\n",
       "      <td>1.220044e-01</td>\n",
       "      <td>0.744984</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.828282e-03</td>\n",
       "      <td>2.916586e-09</td>\n",
       "      <td>8.731569e-04</td>\n",
       "      <td>0.004506</td>\n",
       "      <td>0.435632</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2920991</td>\n",
       "      <td>2.595891e-01</td>\n",
       "      <td>3.598285e-01</td>\n",
       "      <td>2.001547e-01</td>\n",
       "      <td>3.428594e-01</td>\n",
       "      <td>2.200349e-02</td>\n",
       "      <td>2.061667e-01</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>...</td>\n",
       "      <td>1.815006e-01</td>\n",
       "      <td>1.250446e-01</td>\n",
       "      <td>0.744984</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.823206e-03</td>\n",
       "      <td>3.666259e-09</td>\n",
       "      <td>9.606651e-04</td>\n",
       "      <td>0.004506</td>\n",
       "      <td>0.432639</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id rs_id_dbSNP147_GRCh37p13  Adipose_Subcutaneous  \\\n",
       "0  ENSG00000246089                 rs725438          1.043269e-21   \n",
       "1  ENSG00000253893                rs2920992          2.691379e-01   \n",
       "2  ENSG00000253893                rs2979150          2.811242e-01   \n",
       "3  ENSG00000253893                rs2976929          2.671713e-01   \n",
       "4  ENSG00000253893                rs2920991          2.595891e-01   \n",
       "\n",
       "   Adipose_Visceral_Omentum  Adrenal_Gland  Artery_Aorta  Artery_Coronary  \\\n",
       "0              6.992728e-11   1.086137e-16  2.721382e-17     7.937750e-09   \n",
       "1              3.598285e-01   2.005095e-01  3.509573e-01     2.239074e-02   \n",
       "2              3.600600e-01   2.037117e-01  3.421716e-01     2.260352e-02   \n",
       "3              3.598285e-01   2.003048e-01  3.478835e-01     2.228504e-02   \n",
       "4              3.598285e-01   2.001547e-01  3.428594e-01     2.200349e-02   \n",
       "\n",
       "   Artery_Tibial  Brain_Amygdala  Brain_Anterior_cingulate_cortex_BA24  ...  \\\n",
       "0   2.789501e-14               1                              0.003194  ...   \n",
       "1   2.079237e-01               1                              0.000353  ...   \n",
       "2   2.303030e-01               1                              0.000356  ...   \n",
       "3   2.089616e-01               1                              0.000353  ...   \n",
       "4   2.061667e-01               1                              0.000353  ...   \n",
       "\n",
       "   Skin_Not_Sun_Exposed_Suprapubic  Skin_Sun_Exposed_Lower_leg  \\\n",
       "0                     6.476060e-10                4.340546e-15   \n",
       "1                     1.634199e-01                1.208264e-01   \n",
       "2                     1.636915e-01                1.190036e-01   \n",
       "3                     1.634199e-01                1.220044e-01   \n",
       "4                     1.815006e-01                1.250446e-01   \n",
       "\n",
       "   Small_Intestine_Terminal_Ileum        Spleen       Stomach        Testis  \\\n",
       "0                        0.000019  3.087270e-10  2.229960e-08  3.699122e-25   \n",
       "1                        0.744984  1.000000e+00  1.831117e-03  2.916586e-09   \n",
       "2                        0.746993  1.000000e+00  1.853348e-03  2.886361e-09   \n",
       "3                        0.744984  1.000000e+00  1.828282e-03  2.916586e-09   \n",
       "4                        0.744984  1.000000e+00  1.823206e-03  3.666259e-09   \n",
       "\n",
       "        Thyroid    Uterus    Vagina   Whole_Blood  \n",
       "0  2.288615e-13  0.043879  0.000399  9.007175e-25  \n",
       "1  8.773499e-04  0.004506  0.437609  1.000000e+00  \n",
       "2  9.617443e-04  0.003875  0.441878  1.000000e+00  \n",
       "3  8.731569e-04  0.004506  0.435632  1.000000e+00  \n",
       "4  9.606651e-04  0.004506  0.432639  1.000000e+00  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissues_data_frame = {}\n",
    "tissues_data_frame['gene_id'] = []\n",
    "tissues_data_frame[rs_name] = []\n",
    "for t in tissues_names: tissues_data_frame[t] = []\n",
    "for i in pairs_df.index:\n",
    "    gene_stbl_id = pairs_df.loc[i,]['gene_id']\n",
    "    variant_name = pairs_df.loc[i,][rs_name]\n",
    "    if (gene_stbl_id, variant_name) in interaction__tissues_pval:\n",
    "        tissues_data_frame['gene_id'].append(gene_stbl_id)\n",
    "        tissues_data_frame[rs_name].append(variant_name)\n",
    "        for t in tissues_names:\n",
    "            if t in interaction__tissues_pval[(gene_stbl_id, variant_name)]:\n",
    "                tissues_data_frame[t].append(interaction__tissues_pval[(gene_stbl_id, variant_name)][t])\n",
    "            else:\n",
    "                tissues_data_frame[t].append(1)\n",
    "tissues_pval = pd.DataFrame(tissues_data_frame)\n",
    "tissues_pval.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12668 1832\n"
     ]
    }
   ],
   "source": [
    "c = 0\n",
    "n = 0\n",
    "interaction__tissues_beta = {}\n",
    "for i in pairs_df.index:\n",
    "    gene_stbl_id = pairs_df.loc[i,]['gene_id']\n",
    "    variant_name = pairs_df.loc[i,][rs_name]\n",
    "    for tissue in tissues_names:\n",
    "        ext = \"/eqtl/id/homo_sapiens/\"+gene_stbl_id+\"?statistic=beta;variant_name=\"+variant_name+\";tissue=\" + tissue\n",
    "        r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "        if not r.ok:\n",
    "            try:\n",
    "                r.raise_for_status()\n",
    "            except:\n",
    "                n += 1\n",
    "        try:\n",
    "            decoded = r.json()\n",
    "        except:\n",
    "            n += 1\n",
    "        if \"error\" not in repr(decoded) and len(repr(decoded)) > 2: \n",
    "            if (gene_stbl_id, variant_name) not in interaction__tissues_beta:\n",
    "                interaction__tissues_beta[(gene_stbl_id, variant_name)] = {}\n",
    "                interaction__tissues_beta[(gene_stbl_id, variant_name)][tissue] = decoded[0]['value']\n",
    "            else:\n",
    "                interaction__tissues_beta[(gene_stbl_id, variant_name)][tissue] = decoded[0]['value']\n",
    "            c += 1\n",
    "print(c, n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "      <th>Adipose_Subcutaneous</th>\n",
       "      <th>Adipose_Visceral_Omentum</th>\n",
       "      <th>Adrenal_Gland</th>\n",
       "      <th>Artery_Aorta</th>\n",
       "      <th>Artery_Coronary</th>\n",
       "      <th>Artery_Tibial</th>\n",
       "      <th>Brain_Amygdala</th>\n",
       "      <th>Brain_Anterior_cingulate_cortex_BA24</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole_Blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>rs725438</td>\n",
       "      <td>-0.658403</td>\n",
       "      <td>-0.647966</td>\n",
       "      <td>-0.936025</td>\n",
       "      <td>-0.713329</td>\n",
       "      <td>-0.631175</td>\n",
       "      <td>-0.452004</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.464902</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.461236</td>\n",
       "      <td>-0.447576</td>\n",
       "      <td>-0.638269</td>\n",
       "      <td>-0.81166</td>\n",
       "      <td>-0.473882</td>\n",
       "      <td>-0.883826</td>\n",
       "      <td>-0.480686</td>\n",
       "      <td>-0.383610</td>\n",
       "      <td>-0.616397</td>\n",
       "      <td>-0.677099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2920992</td>\n",
       "      <td>-0.097214</td>\n",
       "      <td>0.107471</td>\n",
       "      <td>-0.182481</td>\n",
       "      <td>-0.099121</td>\n",
       "      <td>-0.274940</td>\n",
       "      <td>-0.118052</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.617426</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169422</td>\n",
       "      <td>-0.135644</td>\n",
       "      <td>-0.065092</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320903</td>\n",
       "      <td>-0.737579</td>\n",
       "      <td>-0.292812</td>\n",
       "      <td>-0.598346</td>\n",
       "      <td>-0.118196</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2979150</td>\n",
       "      <td>-0.094421</td>\n",
       "      <td>0.107488</td>\n",
       "      <td>-0.181278</td>\n",
       "      <td>-0.101037</td>\n",
       "      <td>-0.274442</td>\n",
       "      <td>-0.112311</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.618445</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169465</td>\n",
       "      <td>-0.136068</td>\n",
       "      <td>-0.064566</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320723</td>\n",
       "      <td>-0.738005</td>\n",
       "      <td>-0.290298</td>\n",
       "      <td>-0.602307</td>\n",
       "      <td>-0.115607</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2976929</td>\n",
       "      <td>-0.097675</td>\n",
       "      <td>0.107471</td>\n",
       "      <td>-0.182548</td>\n",
       "      <td>-0.099779</td>\n",
       "      <td>-0.275128</td>\n",
       "      <td>-0.117793</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.617426</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169422</td>\n",
       "      <td>-0.135275</td>\n",
       "      <td>-0.065092</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320917</td>\n",
       "      <td>-0.737579</td>\n",
       "      <td>-0.293012</td>\n",
       "      <td>-0.598346</td>\n",
       "      <td>-0.118846</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>rs2920991</td>\n",
       "      <td>-0.099342</td>\n",
       "      <td>0.107471</td>\n",
       "      <td>-0.182616</td>\n",
       "      <td>-0.100867</td>\n",
       "      <td>-0.275712</td>\n",
       "      <td>-0.118531</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.617426</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.164296</td>\n",
       "      <td>-0.134278</td>\n",
       "      <td>-0.065092</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320940</td>\n",
       "      <td>-0.735544</td>\n",
       "      <td>-0.292039</td>\n",
       "      <td>-0.598346</td>\n",
       "      <td>-0.119806</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 49 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene_id rs_id_dbSNP147_GRCh37p13  Adipose_Subcutaneous  \\\n",
       "0  ENSG00000246089                 rs725438             -0.658403   \n",
       "1  ENSG00000253893                rs2920992             -0.097214   \n",
       "2  ENSG00000253893                rs2979150             -0.094421   \n",
       "3  ENSG00000253893                rs2976929             -0.097675   \n",
       "4  ENSG00000253893                rs2920991             -0.099342   \n",
       "\n",
       "   Adipose_Visceral_Omentum  Adrenal_Gland  Artery_Aorta  Artery_Coronary  \\\n",
       "0                 -0.647966      -0.936025     -0.713329        -0.631175   \n",
       "1                  0.107471      -0.182481     -0.099121        -0.274940   \n",
       "2                  0.107488      -0.181278     -0.101037        -0.274442   \n",
       "3                  0.107471      -0.182548     -0.099779        -0.275128   \n",
       "4                  0.107471      -0.182616     -0.100867        -0.275712   \n",
       "\n",
       "   Artery_Tibial  Brain_Amygdala  Brain_Anterior_cingulate_cortex_BA24  ...  \\\n",
       "0      -0.452004               1                             -0.464902  ...   \n",
       "1      -0.118052               1                             -0.617426  ...   \n",
       "2      -0.112311               1                             -0.618445  ...   \n",
       "3      -0.117793               1                             -0.617426  ...   \n",
       "4      -0.118531               1                             -0.617426  ...   \n",
       "\n",
       "   Skin_Not_Sun_Exposed_Suprapubic  Skin_Sun_Exposed_Lower_leg  \\\n",
       "0                        -0.461236                   -0.447576   \n",
       "1                        -0.169422                   -0.135644   \n",
       "2                        -0.169465                   -0.136068   \n",
       "3                        -0.169422                   -0.135275   \n",
       "4                        -0.164296                   -0.134278   \n",
       "\n",
       "   Small_Intestine_Terminal_Ileum   Spleen   Stomach    Testis   Thyroid  \\\n",
       "0                       -0.638269 -0.81166 -0.473882 -0.883826 -0.480686   \n",
       "1                       -0.065092  1.00000 -0.320903 -0.737579 -0.292812   \n",
       "2                       -0.064566  1.00000 -0.320723 -0.738005 -0.290298   \n",
       "3                       -0.065092  1.00000 -0.320917 -0.737579 -0.293012   \n",
       "4                       -0.065092  1.00000 -0.320940 -0.735544 -0.292039   \n",
       "\n",
       "     Uterus    Vagina  Whole_Blood  \n",
       "0 -0.383610 -0.616397    -0.677099  \n",
       "1 -0.598346 -0.118196     1.000000  \n",
       "2 -0.602307 -0.115607     1.000000  \n",
       "3 -0.598346 -0.118846     1.000000  \n",
       "4 -0.598346 -0.119806     1.000000  \n",
       "\n",
       "[5 rows x 49 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissues_data_frame = {}\n",
    "tissues_data_frame['gene_id'] = []\n",
    "tissues_data_frame[rs_name] = []\n",
    "for t in tissues_names: tissues_data_frame[t] = []\n",
    "for i in pairs_df.index:\n",
    "    gene_stbl_id = pairs_df.loc[i,]['gene_id']\n",
    "    variant_name = pairs_df.loc[i,][rs_name]\n",
    "    if (gene_stbl_id, variant_name) in interaction__tissues_beta:\n",
    "        tissues_data_frame['gene_id'].append(gene_stbl_id)\n",
    "        tissues_data_frame[rs_name].append(variant_name)\n",
    "        for t in tissues_names:\n",
    "            if t in interaction__tissues_beta[(gene_stbl_id, variant_name)]:\n",
    "                tissues_data_frame[t].append(interaction__tissues_beta[(gene_stbl_id, variant_name)][t])\n",
    "            else:\n",
    "                tissues_data_frame[t].append(1)\n",
    "tissues_beta = pd.DataFrame(tissues_data_frame)\n",
    "tissues_beta.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'tissues_features_beta' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-30-eb5130099738>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpairs_df\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpairs_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtissues_features_beta\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdrop_duplicates\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'tissues_features_beta' is not defined"
     ]
    }
   ],
   "source": [
    "pairs_df = pairs_df.drop(columns=tissues_features_beta).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues_features_beta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "      <th>...</th>\n",
       "      <th>ma_count</th>\n",
       "      <th>maf</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>slope</th>\n",
       "      <th>slope_se</th>\n",
       "      <th>pval_nominal_threshold</th>\n",
       "      <th>min_pval_nominal</th>\n",
       "      <th>pval_beta</th>\n",
       "      <th>enhancer_bin</th>\n",
       "      <th>promoter_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "      <td>8</td>\n",
       "      <td>6262831</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>rs725438</td>\n",
       "      <td>...</td>\n",
       "      <td>75</td>\n",
       "      <td>0.323276</td>\n",
       "      <td>1.283440e-06</td>\n",
       "      <td>-0.613816</td>\n",
       "      <td>0.118679</td>\n",
       "      <td>0.000005</td>\n",
       "      <td>1.773260e-08</td>\n",
       "      <td>9.862700e-05</td>\n",
       "      <td>6260000</td>\n",
       "      <td>6405000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258448</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920992</td>\n",
       "      <td>...</td>\n",
       "      <td>112</td>\n",
       "      <td>0.478632</td>\n",
       "      <td>7.157010e-06</td>\n",
       "      <td>-0.543850</td>\n",
       "      <td>0.114459</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "      <td>8255000</td>\n",
       "      <td>8225000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258455</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2979150</td>\n",
       "      <td>...</td>\n",
       "      <td>112</td>\n",
       "      <td>0.478632</td>\n",
       "      <td>7.157010e-06</td>\n",
       "      <td>-0.543850</td>\n",
       "      <td>0.114459</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "      <td>8255000</td>\n",
       "      <td>8225000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2976929</td>\n",
       "      <td>...</td>\n",
       "      <td>112</td>\n",
       "      <td>0.478632</td>\n",
       "      <td>7.157010e-06</td>\n",
       "      <td>-0.543850</td>\n",
       "      <td>0.114459</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "      <td>8255000</td>\n",
       "      <td>8225000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8259117</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920991</td>\n",
       "      <td>...</td>\n",
       "      <td>107</td>\n",
       "      <td>0.490826</td>\n",
       "      <td>6.822610e-07</td>\n",
       "      <td>-0.616851</td>\n",
       "      <td>0.115898</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>7.250160e-11</td>\n",
       "      <td>3.037940e-07</td>\n",
       "      <td>8255000</td>\n",
       "      <td>8225000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id          gene_id           enhancer_id  \\\n",
       "0  8_6262831_A_T_b37  ENSG00000246089  chr8:6262801-6263000   \n",
       "1  8_8258448_C_G_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "2  8_8258455_G_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "3  8_8258712_T_C_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "4  8_8259117_T_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "\n",
       "            promoter_id  chr  variant_pos ref alt  num_alt_per_site  \\\n",
       "0  chr8:6405000-6408201    8      6262831   A   T                 1   \n",
       "1  chr8:8226600-8230601    8      8258448   C   G                 1   \n",
       "2  chr8:8226600-8230601    8      8258455   G   A                 1   \n",
       "3  chr8:8226600-8230601    8      8258712   T   C                 1   \n",
       "4  chr8:8226600-8230601    8      8259117   T   A                 1   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  ...  ma_count       maf  pval_nominal     slope  \\\n",
       "0                 rs725438  ...        75  0.323276  1.283440e-06 -0.613816   \n",
       "1                rs2920992  ...       112  0.478632  7.157010e-06 -0.543850   \n",
       "2                rs2979150  ...       112  0.478632  7.157010e-06 -0.543850   \n",
       "3                rs2976929  ...       112  0.478632  7.157010e-06 -0.543850   \n",
       "4                rs2920991  ...       107  0.490826  6.822610e-07 -0.616851   \n",
       "\n",
       "   slope_se  pval_nominal_threshold  min_pval_nominal     pval_beta  \\\n",
       "0  0.118679                0.000005      1.773260e-08  9.862700e-05   \n",
       "1  0.114459                0.000019      7.250160e-11  3.037940e-07   \n",
       "2  0.114459                0.000019      7.250160e-11  3.037940e-07   \n",
       "3  0.114459                0.000019      7.250160e-11  3.037940e-07   \n",
       "4  0.115898                0.000019      7.250160e-11  3.037940e-07   \n",
       "\n",
       "   enhancer_bin  promoter_bin  \n",
       "0       6260000       6405000  \n",
       "1       8255000       8225000  \n",
       "2       8255000       8225000  \n",
       "3       8255000       8225000  \n",
       "4       8255000       8225000  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join all data \n",
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(329, 123)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# enhancers\n",
    "pairs_df = pairs_df.merge(enhancer_linear, right_on = 'enhancer_id', left_on = 'enhancer_id')\n",
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(329, 224)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# promoters\n",
    "pairs_df = pairs_df.merge(promoter_linear, right_on = 'promoter_id', left_on = 'promoter_id')\n",
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "hi_c_dict = hi_c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>hi_c</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            promoter_id           enhancer_id  hi_c         variant_id  \\\n",
       "0  chr8:6405000-6408201  chr8:6262801-6263000   3.0  8_6262831_A_T_b37   \n",
       "1  chr8:8226600-8230601  chr8:8258201-8259600   8.0  8_8258448_C_G_b37   \n",
       "2  chr8:8226600-8230601  chr8:8258201-8259600   8.0  8_8258455_G_A_b37   \n",
       "3  chr8:8226600-8230601  chr8:8258201-8259600   8.0  8_8258712_T_C_b37   \n",
       "4  chr8:8226600-8230601  chr8:8258201-8259600   8.0  8_8259117_T_A_b37   \n",
       "\n",
       "           gene_id  \n",
       "0  ENSG00000246089  \n",
       "1  ENSG00000253893  \n",
       "2  ENSG00000253893  \n",
       "3  ENSG00000253893  \n",
       "4  ENSG00000253893  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hi_c = pd.DataFrame(hi_c_dict)\n",
    "hi_c.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# hi-c\n",
    "pairs_df = pairs_df.merge(hi_c, how='inner', right_on=['promoter_id', 'enhancer_id', 'variant_id', 'gene_id'], left_on=['promoter_id', 'enhancer_id', 'variant_id', 'gene_id'])\n",
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df = pairs_df.drop(columns=tissues_features_beta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# eQTL\n",
    "tissues_beta = tissues_beta.drop_duplicates()\n",
    "tissues_beta = tissues_beta.fillna(1)\n",
    "pairs_df = pairs_df.merge(tissues_beta, how='left', right_on=['gene_id', rs_name], left_on=['gene_id', rs_name])\n",
    "columns = []\n",
    "for i in pairs_df.columns:\n",
    "    if i in tissues_names: columns.append(i+'_beta')\n",
    "    else: columns.append(i)\n",
    "pairs_df.columns = columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic_beta</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg_beta</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum_beta</th>\n",
       "      <th>Spleen_beta</th>\n",
       "      <th>Stomach_beta</th>\n",
       "      <th>Testis_beta</th>\n",
       "      <th>Thyroid_beta</th>\n",
       "      <th>Uterus_beta</th>\n",
       "      <th>Vagina_beta</th>\n",
       "      <th>Whole_Blood_beta</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "      <td>8</td>\n",
       "      <td>6262831</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>rs725438</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.461236</td>\n",
       "      <td>-0.447576</td>\n",
       "      <td>-0.638269</td>\n",
       "      <td>-0.81166</td>\n",
       "      <td>-0.473882</td>\n",
       "      <td>-0.883826</td>\n",
       "      <td>-0.480686</td>\n",
       "      <td>-0.383610</td>\n",
       "      <td>-0.616397</td>\n",
       "      <td>-0.677099</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258448</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920992</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169422</td>\n",
       "      <td>-0.135644</td>\n",
       "      <td>-0.065092</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320903</td>\n",
       "      <td>-0.737579</td>\n",
       "      <td>-0.292812</td>\n",
       "      <td>-0.598346</td>\n",
       "      <td>-0.118196</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258455</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2979150</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169465</td>\n",
       "      <td>-0.136068</td>\n",
       "      <td>-0.064566</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320723</td>\n",
       "      <td>-0.738005</td>\n",
       "      <td>-0.290298</td>\n",
       "      <td>-0.602307</td>\n",
       "      <td>-0.115607</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2976929</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.169422</td>\n",
       "      <td>-0.135275</td>\n",
       "      <td>-0.065092</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320917</td>\n",
       "      <td>-0.737579</td>\n",
       "      <td>-0.293012</td>\n",
       "      <td>-0.598346</td>\n",
       "      <td>-0.118846</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8259117</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920991</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.164296</td>\n",
       "      <td>-0.134278</td>\n",
       "      <td>-0.065092</td>\n",
       "      <td>1.00000</td>\n",
       "      <td>-0.320940</td>\n",
       "      <td>-0.735544</td>\n",
       "      <td>-0.292039</td>\n",
       "      <td>-0.598346</td>\n",
       "      <td>-0.119806</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 271 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id          gene_id           enhancer_id  \\\n",
       "0  8_6262831_A_T_b37  ENSG00000246089  chr8:6262801-6263000   \n",
       "1  8_8258448_C_G_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "2  8_8258455_G_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "3  8_8258712_T_C_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "4  8_8259117_T_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "\n",
       "            promoter_id  chr  variant_pos ref alt  num_alt_per_site  \\\n",
       "0  chr8:6405000-6408201    8      6262831   A   T                 1   \n",
       "1  chr8:8226600-8230601    8      8258448   C   G                 1   \n",
       "2  chr8:8226600-8230601    8      8258455   G   A                 1   \n",
       "3  chr8:8226600-8230601    8      8258712   T   C                 1   \n",
       "4  chr8:8226600-8230601    8      8259117   T   A                 1   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  ...  Skin_Not_Sun_Exposed_Suprapubic_beta  \\\n",
       "0                 rs725438  ...                             -0.461236   \n",
       "1                rs2920992  ...                             -0.169422   \n",
       "2                rs2979150  ...                             -0.169465   \n",
       "3                rs2976929  ...                             -0.169422   \n",
       "4                rs2920991  ...                             -0.164296   \n",
       "\n",
       "   Skin_Sun_Exposed_Lower_leg_beta  Small_Intestine_Terminal_Ileum_beta  \\\n",
       "0                        -0.447576                            -0.638269   \n",
       "1                        -0.135644                            -0.065092   \n",
       "2                        -0.136068                            -0.064566   \n",
       "3                        -0.135275                            -0.065092   \n",
       "4                        -0.134278                            -0.065092   \n",
       "\n",
       "   Spleen_beta  Stomach_beta  Testis_beta  Thyroid_beta  Uterus_beta  \\\n",
       "0     -0.81166     -0.473882    -0.883826     -0.480686    -0.383610   \n",
       "1      1.00000     -0.320903    -0.737579     -0.292812    -0.598346   \n",
       "2      1.00000     -0.320723    -0.738005     -0.290298    -0.602307   \n",
       "3      1.00000     -0.320917    -0.737579     -0.293012    -0.598346   \n",
       "4      1.00000     -0.320940    -0.735544     -0.292039    -0.598346   \n",
       "\n",
       "   Vagina_beta  Whole_Blood_beta  \n",
       "0    -0.616397         -0.677099  \n",
       "1    -0.118196          1.000000  \n",
       "2    -0.115607          1.000000  \n",
       "3    -0.118846          1.000000  \n",
       "4    -0.119806          1.000000  \n",
       "\n",
       "[5 rows x 271 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues_pval = tissues_pval.drop_duplicates()\n",
    "tissues_pval = tissues_pval.fillna(1)\n",
    "pairs_df = pairs_df.merge(tissues_pval, how='left', right_on=['gene_id', rs_name], left_on=['gene_id', rs_name])\n",
    "columns = []\n",
    "for i in pairs_df.columns:\n",
    "    if i in tissues_names: columns.append(i+'_pval')\n",
    "    else: columns.append(i)\n",
    "pairs_df.columns = columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(290, 412)"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df = pairs_df.drop_duplicates()\n",
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "pairs_df.to_csv('../data/chromosome8_all_data_pval.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add Z-Scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>variant_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>enhancer_id</th>\n",
       "      <th>promoter_id</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_alt_per_site</th>\n",
       "      <th>rs_id_dbSNP147_GRCh37p13</th>\n",
       "      <th>...</th>\n",
       "      <th>Skin_Not_Sun_Exposed_Suprapubic</th>\n",
       "      <th>Skin_Sun_Exposed_Lower_leg</th>\n",
       "      <th>Small_Intestine_Terminal_Ileum</th>\n",
       "      <th>Spleen</th>\n",
       "      <th>Stomach</th>\n",
       "      <th>Testis</th>\n",
       "      <th>Thyroid</th>\n",
       "      <th>Uterus</th>\n",
       "      <th>Vagina</th>\n",
       "      <th>Whole_Blood</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8_6262831_A_T_b37</td>\n",
       "      <td>ENSG00000246089</td>\n",
       "      <td>chr8:6262801-6263000</td>\n",
       "      <td>chr8:6405000-6408201</td>\n",
       "      <td>8</td>\n",
       "      <td>6262831</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "      <td>rs725438</td>\n",
       "      <td>...</td>\n",
       "      <td>6.476060e-10</td>\n",
       "      <td>4.340546e-15</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>3.087270e-10</td>\n",
       "      <td>2.229960e-08</td>\n",
       "      <td>3.699122e-25</td>\n",
       "      <td>2.288615e-13</td>\n",
       "      <td>0.043879</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>9.007175e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8_8258448_C_G_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258448</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920992</td>\n",
       "      <td>...</td>\n",
       "      <td>1.634199e-01</td>\n",
       "      <td>1.208264e-01</td>\n",
       "      <td>0.744984</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.831117e-03</td>\n",
       "      <td>2.916586e-09</td>\n",
       "      <td>8.773499e-04</td>\n",
       "      <td>0.004506</td>\n",
       "      <td>0.437609</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8_8258455_G_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258455</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2979150</td>\n",
       "      <td>...</td>\n",
       "      <td>1.636915e-01</td>\n",
       "      <td>1.190036e-01</td>\n",
       "      <td>0.746993</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.853348e-03</td>\n",
       "      <td>2.886361e-09</td>\n",
       "      <td>9.617443e-04</td>\n",
       "      <td>0.003875</td>\n",
       "      <td>0.441878</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>8_8258712_T_C_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8258712</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2976929</td>\n",
       "      <td>...</td>\n",
       "      <td>1.634199e-01</td>\n",
       "      <td>1.220044e-01</td>\n",
       "      <td>0.744984</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.828282e-03</td>\n",
       "      <td>2.916586e-09</td>\n",
       "      <td>8.731569e-04</td>\n",
       "      <td>0.004506</td>\n",
       "      <td>0.435632</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8_8259117_T_A_b37</td>\n",
       "      <td>ENSG00000253893</td>\n",
       "      <td>chr8:8258201-8259600</td>\n",
       "      <td>chr8:8226600-8230601</td>\n",
       "      <td>8</td>\n",
       "      <td>8259117</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>rs2920991</td>\n",
       "      <td>...</td>\n",
       "      <td>1.815006e-01</td>\n",
       "      <td>1.250446e-01</td>\n",
       "      <td>0.744984</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.823206e-03</td>\n",
       "      <td>3.666259e-09</td>\n",
       "      <td>9.606651e-04</td>\n",
       "      <td>0.004506</td>\n",
       "      <td>0.432639</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 272 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          variant_id          gene_id           enhancer_id  \\\n",
       "0  8_6262831_A_T_b37  ENSG00000246089  chr8:6262801-6263000   \n",
       "1  8_8258448_C_G_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "2  8_8258455_G_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "3  8_8258712_T_C_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "4  8_8259117_T_A_b37  ENSG00000253893  chr8:8258201-8259600   \n",
       "\n",
       "            promoter_id  chr  variant_pos ref alt  num_alt_per_site  \\\n",
       "0  chr8:6405000-6408201    8      6262831   A   T                 1   \n",
       "1  chr8:8226600-8230601    8      8258448   C   G                 1   \n",
       "2  chr8:8226600-8230601    8      8258455   G   A                 1   \n",
       "3  chr8:8226600-8230601    8      8258712   T   C                 1   \n",
       "4  chr8:8226600-8230601    8      8259117   T   A                 1   \n",
       "\n",
       "  rs_id_dbSNP147_GRCh37p13  ...  Skin_Not_Sun_Exposed_Suprapubic  \\\n",
       "0                 rs725438  ...                     6.476060e-10   \n",
       "1                rs2920992  ...                     1.634199e-01   \n",
       "2                rs2979150  ...                     1.636915e-01   \n",
       "3                rs2976929  ...                     1.634199e-01   \n",
       "4                rs2920991  ...                     1.815006e-01   \n",
       "\n",
       "   Skin_Sun_Exposed_Lower_leg  Small_Intestine_Terminal_Ileum        Spleen  \\\n",
       "0                4.340546e-15                        0.000019  3.087270e-10   \n",
       "1                1.208264e-01                        0.744984  1.000000e+00   \n",
       "2                1.190036e-01                        0.746993  1.000000e+00   \n",
       "3                1.220044e-01                        0.744984  1.000000e+00   \n",
       "4                1.250446e-01                        0.744984  1.000000e+00   \n",
       "\n",
       "        Stomach        Testis       Thyroid    Uterus    Vagina   Whole_Blood  \n",
       "0  2.229960e-08  3.699122e-25  2.288615e-13  0.043879  0.000399  9.007175e-25  \n",
       "1  1.831117e-03  2.916586e-09  8.773499e-04  0.004506  0.437609  1.000000e+00  \n",
       "2  1.853348e-03  2.886361e-09  9.617443e-04  0.003875  0.441878  1.000000e+00  \n",
       "3  1.828282e-03  2.916586e-09  8.731569e-04  0.004506  0.435632  1.000000e+00  \n",
       "4  1.823206e-03  3.666259e-09  9.606651e-04  0.004506  0.432639  1.000000e+00  \n",
       "\n",
       "[5 rows x 272 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df = pd.read_csv('../data/chromosome8_all_data_pval.csv')\n",
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df['rsid'] = pairs_df[rs_name] + '_' + pairs_df['promoter_id'] + pairs_df['enhancer_id'] \n",
    "pairs_df = pairs_df.drop_duplicates(subset='rsid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(248, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rsid</th>\n",
       "      <th>beta</th>\n",
       "      <th>se</th>\n",
       "      <th>z</th>\n",
       "      <th>prob</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs2639898_chr8:70667200-70670001chr8:71918401-...</td>\n",
       "      <td>0.774308</td>\n",
       "      <td>0.099723</td>\n",
       "      <td>7.76456</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs2923444_chr8:42537600-42545401chr8:42396801-...</td>\n",
       "      <td>-0.634540</td>\n",
       "      <td>0.060890</td>\n",
       "      <td>-10.42110</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1441438_chr8:78665200-78668001chr8:79641601-...</td>\n",
       "      <td>-0.991622</td>\n",
       "      <td>0.091605</td>\n",
       "      <td>-10.82500</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1441438_chr8:78668200-78670000chr8:79641601-...</td>\n",
       "      <td>-0.991622</td>\n",
       "      <td>0.091605</td>\n",
       "      <td>-10.82500</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs72669109_chr8:109362200-109363201chr8:110523...</td>\n",
       "      <td>0.633513</td>\n",
       "      <td>0.142989</td>\n",
       "      <td>4.43050</td>\n",
       "      <td>0.333333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                rsid      beta        se  \\\n",
       "0  rs2639898_chr8:70667200-70670001chr8:71918401-...  0.774308  0.099723   \n",
       "1  rs2923444_chr8:42537600-42545401chr8:42396801-... -0.634540  0.060890   \n",
       "2  rs1441438_chr8:78665200-78668001chr8:79641601-... -0.991622  0.091605   \n",
       "3  rs1441438_chr8:78668200-78670000chr8:79641601-... -0.991622  0.091605   \n",
       "4  rs72669109_chr8:109362200-109363201chr8:110523...  0.633513  0.142989   \n",
       "\n",
       "          z      prob  \n",
       "0   7.76456  1.000000  \n",
       "1 -10.42110  1.000000  \n",
       "2 -10.82500  0.500000  \n",
       "3 -10.82500  0.500000  \n",
       "4   4.43050  0.333333  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_z = pd.read_csv('../data/chromosome8_to_run_finemap.snp', sep=' ')[['rsid', 'beta', 'se', 'z', 'prob']]\n",
    "print(data_z.shape)\n",
    "data_z.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df = pairs_df.merge(data_z, how='inner', left_on='rsid', right_on='rsid')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(248, 417)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pairs_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "enhancer_features = []\n",
    "promoter_features = []\n",
    "\n",
    "for f in pairs_df.columns:\n",
    "    if   f.startswith('enhancer_') and f not in {'enhancer_id', 'enhancer_bin'}: enhancer_features.append(f)\n",
    "    elif f.startswith('promoter_') and f not in {'promoter_id', 'promoter_bin'}: promoter_features.append(f)\n",
    "        \n",
    "tissues_features_pval = list(i for i in pairs_df.columns if i.endswith('_pval'))\n",
    "tissues_features_beta = list(i for i in pairs_df.columns if i.endswith('_beta'))\n",
    "\n",
    "hi_c_features = ['hi_c']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tissues_features_beta = list(i+'_beta' for i in tissues_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_line = pairs_df[enhancer_features+promoter_features].values\n",
    "x_hi_c = pairs_df[hi_c_features].values\n",
    "x_same = pairs_df[hi_c_features+enhancer_features+promoter_features].values\n",
    "x_pval = pairs_df[tissues_features_pval].values\n",
    "x_beta = pairs_df[tissues_features_beta].values\n",
    "x_allb = pairs_df[hi_c_features+enhancer_features+promoter_features+tissues_features_beta].values\n",
    "y = pairs_df['z'].values\n",
    "y = pairs_df['z'].values\n",
    "\n",
    "_, e_line, _, _, _ = methodName.train_random_forest(x_line, y)\n",
    "_, e_hi_c, _, _, _ = methodName.train_random_forest(x_hi_c, y)\n",
    "_, e_same, _, _, _ = methodName.train_random_forest(x_same, y)\n",
    "_, e_pval, _, _, _ = methodName.train_random_forest(x_pval, y)\n",
    "_, e_beta, _, _, _ = methodName.train_random_forest(x_beta, y)\n",
    "_, e_allb, _, _, _ = methodName.train_random_forest(x_allb, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "e_line, e_hi_c, e_same, e_pval, e_beta, e_allb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairs_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Test idea residuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import mean_squared_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "X = pairs_df[tissues_features_beta+enhancer_features+promoter_features]\n",
    "y = pairs_df['z']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "X_internal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_errors = []\n",
    "resd_errors = []\n",
    "eqtl_errors = []\n",
    "for i in range(1000):\n",
    "    \n",
    "    X_internal, X_external, y_internal, y_external = train_test_split(X, y, test_size=0.5)\n",
    "\n",
    "    # Models\n",
    "    model_eqtl = RandomForestRegressor()\n",
    "    model_resd = RandomForestRegressor()\n",
    "    model_full = RandomForestRegressor()\n",
    "\n",
    "    # Train eqtl model\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X_internal, y_internal, test_size=0.2, random_state=12)\n",
    "    model_eqtl = model_eqtl.fit(X_train[tissues_features_beta], y_train)\n",
    "    y_pred = model_eqtl.predict(X_test[tissues_features_beta])\n",
    "    e_test = mean_squared_error(y_test, y_pred)\n",
    "    eqtl_errors.append(e_test)\n",
    "    eqtl_pred = model_eqtl.predict(X_internal[tissues_features_beta])\n",
    "    eqtl_resd = eqtl_pred - y_internal\n",
    "\n",
    "    # Train residual models\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X_internal, eqtl_resd, test_size=0.2, random_state=12)\n",
    "    model_resd = model_resd.fit(X_train[enhancer_features+promoter_features], y_train)\n",
    "    y_pred = model_resd.predict(X_test[enhancer_features+promoter_features])\n",
    "    e_test = mean_squared_error(y_test, y_pred)\n",
    "\n",
    "    # Train full model\n",
    "    X_train, X_test, y_train, y_test = train_test_split(X_internal, y_internal, test_size=0.2, random_state=12)\n",
    "    model_full = model_full.fit(X_train[tissues_features_beta+enhancer_features+promoter_features], y_train)\n",
    "    y_pred = model_full.predict(X_test[tissues_features_beta+enhancer_features+promoter_features])\n",
    "    e_test = mean_squared_error(y_test, y_pred)\n",
    "\n",
    "\n",
    "    # Compare external predictions\n",
    "    external_pred = model_eqtl.predict(X_external[tissues_features_beta])\n",
    "    external_pred += model_resd.predict(X_external[enhancer_features+promoter_features])\n",
    "    e_test = mean_squared_error(y_external, external_pred)\n",
    "    resd_errors.append(e_test)\n",
    "\n",
    "    external_pred = model_full.predict(X_external[tissues_features_beta+enhancer_features+promoter_features])\n",
    "    e_test = mean_squared_error(y_external, external_pred)\n",
    "    full_errors.append(e_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_errors = np.array(full_errors)\n",
    "resd_errors = np.array(resd_errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "74"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(full_errors-resd_errors < 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(3-4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(e_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "print(e_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000254207.1\n",
      "ENSG00000253982.1\n",
      "ENSG00000036448.5\n",
      "ENSG00000246089.3\n",
      "ENSG00000271743.1\n",
      "ENSG00000253893.2\n",
      "ENSG00000269918.1\n",
      "ENSG00000255310.2\n",
      "ENSG00000154319.10\n",
      "ENSG00000136573.8\n",
      "ENSG00000154328.11\n",
      "ENSG00000186523.10\n",
      "ENSG00000255556.2\n",
      "ENSG00000145002.8\n",
      "ENSG00000129422.9\n",
      "ENSG00000158863.17\n",
      "ENSG00000173566.9\n",
      "ENSG00000120910.10\n",
      "ENSG00000254272.1\n",
      "ENSG00000158941.12\n",
      "ENSG00000248738.2\n",
      "ENSG00000120889.8\n",
      "ENSG00000253616.1\n",
      "ENSG00000228451.3\n",
      "ENSG00000104228.8\n",
      "ENSG00000147419.12\n",
      "ENSG00000246339.4\n",
      "ENSG00000104671.3\n",
      "ENSG00000253457.1\n",
      "ENSG00000272338.1\n",
      "ENSG00000172728.11\n",
      "ENSG00000129696.8\n",
      "ENSG00000158669.7\n",
      "ENSG00000070718.7\n",
      "ENSG00000176209.7\n",
      "ENSG00000168172.4\n",
      "ENSG00000120925.9\n",
      "ENSG00000104738.12\n",
      "ENSG00000253140.1\n",
      "ENSG00000168300.9\n",
      "ENSG00000206579.7\n",
      "ENSG00000137574.6\n",
      "ENSG00000169122.7\n",
      "ENSG00000215114.3\n",
      "ENSG00000104388.10\n",
      "ENSG00000254802.1\n",
      "ENSG00000177182.6\n",
      "ENSG00000172817.3\n",
      "ENSG00000104442.5\n",
      "ENSG00000066855.11\n",
      "ENSG00000255107.1\n",
      "ENSG00000221947.3\n",
      "ENSG00000121039.5\n",
      "ENSG00000171033.8\n",
      "ENSG00000104427.7\n",
      "ENSG00000254352.1\n",
      "ENSG00000253919.1\n",
      "ENSG00000133731.5\n",
      "ENSG00000164695.4\n",
      "ENSG00000104497.9\n",
      "ENSG00000133739.11\n",
      "ENSG00000254088.1\n",
      "ENSG00000176623.7\n",
      "ENSG00000251136.4\n",
      "ENSG00000104320.9\n",
      "ENSG00000123119.7\n",
      "ENSG00000253250.2\n",
      "ENSG00000164944.7\n",
      "ENSG00000253704.1\n",
      "ENSG00000164941.9\n",
      "ENSG00000245080.4\n",
      "ENSG00000254248.1\n",
      "ENSG00000175895.3\n",
      "ENSG00000104361.5\n",
      "ENSG00000156509.9\n",
      "ENSG00000164924.13\n",
      "ENSG00000253328.2\n",
      "ENSG00000104517.8\n",
      "ENSG00000253263.1\n",
      "ENSG00000205038.7\n",
      "ENSG00000177570.9\n",
      "ENSG00000136982.5\n",
      "ENSG00000187955.7\n",
      "ENSG00000172172.3\n",
      "ENSG00000172167.3\n",
      "ENSG00000250290.1\n",
      "ENSG00000156787.12\n",
      "ENSG00000189376.7\n",
      "ENSG00000156795.2\n",
      "ENSG00000271975.1\n",
      "ENSG00000183665.4\n",
      "ENSG00000165071.10\n",
      "ENSG00000008513.10\n",
      "ENSG00000167632.10\n",
      "ENSG00000253210.1\n",
      "ENSG00000244998.1\n",
      "ENSG00000171045.10\n",
      "ENSG00000130193.7\n",
      "ENSG00000247317.3\n",
      "ENSG00000185730.3\n",
      "ENSG00000184428.8\n",
      "ENSG00000014164.6\n",
      "ENSG00000254859.1\n",
      "ENSG00000104518.6\n",
      "ENSG00000255050.1\n",
      "ENSG00000147813.11\n",
      "ENSG00000254812.1\n",
      "ENSG00000181097.5\n",
      "ENSG00000178209.10\n",
      "ENSG00000179698.9\n",
      "ENSG00000235173.2\n",
      "ENSG00000185000.5\n",
      "ENSG00000071894.10\n",
      "ENSG00000147804.5\n",
      "ENSG00000160948.9\n",
      "ENSG00000160949.12\n",
      "ENSG00000160972.5\n",
      "ENSG00000167701.9\n",
      "ENSG00000255182.1\n",
      "ENSG00000160959.3\n",
      "ENSG00000196378.7\n",
      "ENSG00000161016.11\n",
      "ENSG00000170631.10\n",
      "ENSG00000182307.8\n"
     ]
    }
   ],
   "source": [
    "for i in eQTL['gene_id'].unique(): print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
